{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "12142ade",
   "metadata": {},
   "source": [
    "The initial code is specifying details abou the authors and then providing information on the pipeline purpose"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bf29251",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/python\n",
    "__author__\t\t= \"Sander Granneman\"\n",
    "__copyright__\t= \"Copyright 2020\"\n",
    "__version__\t\t= \"0.7.0\"\n",
    "__credits__\t\t= [\"Sander Granneman\", \"Edward Wallace\"]\n",
    "__email__\t\t= [\"sgrannem@staffmail.ed.ac.uk\", \"edward.wallace@ed.ac.uk\"]\n",
    "__status__\t\t= \"beta\"\n",
    "\n",
    "\"\"\" CRAC_pipeline_SE_demult_dedup.py\n",
    "\n",
    "This is a pipeline for processing single-end multiplexed CRAC sequencing data.\n",
    "It is forked from CRAC_pipeline_SE.py v0.6.0, also written in ruffus.\n",
    "This pipeline is specialised for single-end data that is multiplexed in-line, so\n",
    "requires trimming of 3' adapters and demultiplexing, including deduplication. This \n",
    "simplifies the logic of the pipeline, no conditional execution needed.\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47b370a9",
   "metadata": {},
   "source": [
    "This section just imports relevant tools to run the pipeline, including from rufus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab76b8c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from ruffus import *\n",
    "from ruffus.cmdline import MESSAGE\n",
    "from collections import defaultdict\n",
    "import ruffus.cmdline as cmdline\n",
    "import subprocess\n",
    "import platform\n",
    "import glob\n",
    "import re\n",
    "import os\n",
    "import argparse\n",
    "import sys\n",
    "import time"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f426f56",
   "metadata": {},
   "source": [
    "This defines the fields of interest that exist in the GFF file, which will allow this informaton to be called later. The general file format (GFF) is used for numerous inputs in 'input_annotation' - inc mRNA transcript locations mapped to the genome, transcript annotations with ncRNAs and mRNAs broken down into parts (UTR, ORF etc) and a set of annotations of transcript features. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3296325",
   "metadata": {},
   "outputs": [],
   "source": [
    "GFF_FIELDS = [\"seqname\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attributes\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af93b91c",
   "metadata": {},
   "source": [
    "Data parsing is a method to strucute data, it takes input data (in human friendly file forms) and breaks it into small segments more usable by computers, often organised in a parse tree. It is a processing library for arguments that are added. \n",
    "\n",
    "Supported actions include storing the argument (singly, or as part of a list), storing a constant value when the argument is encountered (including special handling for true/false values for boolean switches)and counting the number of times an argument is seen.\n",
    "\n",
    "Arguments here are:\n",
    "* The paths to .fastq files, paths to the gff files of annotations, to the chromosome file and the novoalign file (this is everything to aline reads to the genome)\n",
    "* Path to the barcodes file needed for demultiplexing, file of the 3' adapter sequences for trimming, an input variable of the number of mismatches allowed for demultiplexing and how many processors you want to use for this.\n",
    "* The path to a genelist file for pile up analysis and a genome sequence file to use for this.\n",
    "* The gff file containing transcript sequences for count anlaysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f265379",
   "metadata": {},
   "outputs": [],
   "source": [
    "parser = cmdline.get_argparse(description=\"CRAC pipeline for processing single-end multiplexed CRAC data\")\n",
    "parser.add_argument(\"-f\",dest=\"forwardreads\",help=\"the path to your fastq read files.\",metavar=\"data_1.fastq data_2.fastq ...\",nargs=\"*\",default=None)\n",
    "parser.add_argument(\"-g\",\"--gtf\",dest=\"gtf\",help=\"the path to your gtf annotation file\",metavar=\"rRNA.gtf\",default=None)\n",
    "parser.add_argument(\"-c\",\"--chromosome\",dest=\"chromosome\",help=\"the path to your chromosome length file\",metavar=\"chromosome_lengths.txt\",default=None)\n",
    "parser.add_argument(\"--novoindex\",dest=\"novoindex\",help=\"the path to your novoindex file\",metavar=\"yeast.novoindex\",default=None)\n",
    "parser.add_argument(\"--name\",dest=\"name\",help=\"provide a single word describing the run. Default is 'analysis' with a time stamp\",default=\"analysis_%s\" % time.strftime(\"%d%m%Y\"))\n",
    "parser.add_argument(\"--sgr\",dest=\"sgr\",help=\"to make sgr files with read counts for each genomic position. Default is off\",action=\"store_true\",default=False)\n",
    "parser.add_argument(\"-b\",\"--barcodes\",dest=\"barcodes\",help=\"the path to the file containing the list of barcodes. If you do not provide a barcode file, the demultiplexing step will fail\",metavar=\"barcodes.txt\",default=None)\n",
    "parser.add_argument(\"-a\",\"--adapterfile\",dest=\"adapterfile\",help=\"file containing the 3' adapter sequences for trimming the reads using flexbar. If you do not provide adapter file or preset\",default=None)\n",
    "parser.add_argument(\"-aa\",\"--adapterpreset\",dest=\"adapterpreset\",help=\"adapter preset string [TruSeq, SmallRNA, Methyl, Ribo, Nextera, and NexteraMP]. Requires Flexbar version 3.4.0 or later.\",default=None)\n",
    "parser.add_argument(\"-m\",\"--mismatches\",dest=\"mismatches\",type=int,help=\"indicate how many mismatches you allow for demultiplexing. Default is 1\",default=1)\n",
    "parser.add_argument(\"-p\",\"--processors\",dest=\"processors\",type=int,help=\"indicate how many processors you want to use for analyses. Default is 8\",default=8)\n",
    "parser.add_argument(\"--genelist\",dest=\"genelist\",help=\"provide a genelist file for pileup analysis, with one gene name per line,\",default=None)\n",
    "parser.add_argument(\"--genometab\",dest=\"genometab\",help=\"provide a genome sequence file in pyCRAC tab format for pileup analysis\",default=None)\n",
    "parser.add_argument(\"--transcriptgff\",dest=\"transcriptgff\",help=\"provide a gff file containing only transcript sequences for count analysis\",default=None)\n",
    "args = parser.parse_args()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33adc234",
   "metadata": {},
   "source": [
    "In python to define a function the syntax is __def__ function name(parameters): and then on a new line a defintion of the function so the below creates the function getBarcodeInfo that operates on the input parameter BarcodeInfo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df00747f",
   "metadata": {},
   "source": [
    "This set of code is used to define all the functions to be used in the pipeline - these functions can then be provided to rufus alongside their connection and the programme will direct data correctly through this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "266b0328",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this is a function to return the .fastq barcode file in the desired format\n",
    "def getBarcodeInfo(barcodefile):\n",
    "\treturn [\"%s.fastq\" % \"_\".join(line.strip().split()) for line in open(barcodefile,\"r\").readlines()]\n",
    "\t"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "225004db",
   "metadata": {},
   "source": [
    "The program Flexbar preprocesses high-throughput sequencing data efficiently. It demultiplexes barcoded runs and removes adapter sequences - it is available on GitHub"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33728b3d",
   "metadata": {},
   "source": [
    "A command line interpreter allows the user to interact with a program using commands in the form of text lines - users can type their requests and queries as required and obtain the desired results. cmd is a module to provide the line oriented interpreter framework.\n",
    "\n",
    "The os module provides a portable way of using operating system dependent functionality, this is a program that acts as an interface between the system hardware and the user."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d8665f6",
   "metadata": {},
   "source": [
    "NOTE: These functions now all specify inputfile as the input, and outputfile as the output which allows them all to be linked to one another as a file ouput from the first can move into the next."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e5c878e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function defines the input,runs Flexbar on this and defines and output format to remove adapter sequences\n",
    "#qflags is a kind of enum, a data type consisting of a set of named values called elements. \n",
    "#The QFlags class provides a type-safe way of storing OR-combinations of enum values\n",
    "\n",
    "def runFlexBar(inputfile,outputfile,qflags = \"-q TAIL -qf i1.8 -qt 30 --qtrim-post-removal\"):\n",
    "\t\"\"\" runs Flexbar on the data to remove the adapter sequence from the forward reads \"\"\"\n",
    "\tif args.adapterfile:\n",
    "\t\tcmd = \"flexbar -r '%s' --output-reads '%s' -n 10 %s -ao 5 --adapters '%s'\" % (inputfile,outputfile,qflags,args.adapterfile)\n",
    "\telif args.adapterpreset:\n",
    "\t\tcmd = \"flexbar -r '%s' --output-reads '%s' -n 10 '%s' -ao 5 -aa '%s'\" % (inputfile,outputfile,qflags,args.adapterpreset)\n",
    "\telse:\n",
    "\t\tcmd = \"flexbar -r '%s' --output-reads '%s' -n 10 '%s' -ao 5\" % (inputfile,outputfile,qflags)\n",
    "\tlogger.info(cmd)\n",
    "\tos.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "baf6d10d",
   "metadata": {},
   "source": [
    "pyBarcodeFilter.py is one of the codes provided in the pyCRAC software - The tool looks for barcodes in 5' ends of reads and generates separate data files for each barcode in a minimal FASTQ format. Once the tool identies a barcode, the barcode sequence and corresponding quality characters are removed from the raw data and placed in a separate output file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "999c9da6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this function used the Barcodefilter tool to demultiplex the input .fastq file with all the raw reads\n",
    "#notice here that -f, -b, -m etc are arguments specified in the parsing at the beginning, so rufus can call what to use for the command\n",
    "def demultiplexSamples(inputfile,outputfile):\n",
    "\t\"\"\" demultiplexes all the samples \"\"\"\n",
    "\tos.chdir(os.path.join(root_dir,\"demultiplexed\"))\n",
    "\tcmd = \"pyBarcodeFilter.py -f '%s' -b '%s' -m '%s'\" % (inputfile,os.path.join(home_dir,args.barcodes),args.mismatches)\n",
    "\tlogger.info(cmd)\n",
    "\tos.system(cmd)\n",
    "\tos.chdir(root_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "515243e9",
   "metadata": {},
   "source": [
    "During demultiplexing, pyBarcodeFilter will attach the two hashes and random barcode sequence to the read header. pyFastqDuplicateRemover scans the header for the presence of two hashes near the end and assumes that the sequence following these hashes is the random barcode sequence. If it encounters two identical (paired-) sequences with the same random barcode sequences, it assumes they are PCR duplicates and collapse them into one sequence. Produces a fasta output file\n",
    "\n",
    "e.g FCC0TU2ACXX:4:1101:1968:2135#ACAGTGAT1##GTTCTC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2ab8139",
   "metadata": {},
   "outputs": [],
   "source": [
    "def collapseDuplicates(inputfile,outputfile):\n",
    "\t\"\"\" Removes the last random nucleotide from the forward read, collapses the data and then splits it again into fasta files \"\"\"\n",
    "\tcmd = \"pyFastqDuplicateRemover.py -f '%s' -o '%s'\" % (inputfile,outputfile)\n",
    "\tlogger.info(cmd)\n",
    "\tos.system(cmd)\n",
    "\t"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3304da9b",
   "metadata": {},
   "source": [
    "Novoalign runs as a package in python, it is an alignment tool for Illumina reads and accurately maps reads to a reference genome.\n",
    "* -d specifies the reference genome for novoalign\n",
    "* -o specifies the output format\n",
    "* -f was specified as the forward reads in parsing\n",
    "* -r sets the rules for handling of reads with multiple alignment locations, Random mode will randomly chose a single alignment location from amongst all the alignment results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e93c6e65",
   "metadata": {},
   "outputs": [],
   "source": [
    "def alignReads(inputfile,outputfile,alignfiletype = \"SAM\"):\n",
    "\t\"\"\" Runs novoalign on all the collapsed files\"\"\"\n",
    "\tcmd = \"novoalign -d '%s' -o '%s' -f '%s' -r Random > '%s'\" % (args.novoindex,alignfiletype,inputfile,outputfile)\n",
    "\tlogger.info(cmd)\n",
    "\tos.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76a4dc86",
   "metadata": {},
   "source": [
    "After aligning the reads to the reference genome the  rst thing we routinely do is generate hit tables describing the number of reads that mapped to genes or transcripts and to which features they were mapped. The pyReadCounters program takes the Novoalign or BAM/SAM file and asks which reads overlap with genomic features from a GTF file. This produces three outputs: \n",
    "* (1) A text file containing information about the complexity of the sequencing data\n",
    "* (2) a GTF file containing all the genomic mapping positions of cDNA sequences and overlapping genomic features (both sense and anti -sense)\n",
    "* (3) a hit table text file that shows for each genomic feature the number of intervals that mapped sense or anti -sense."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "465315d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def runPyReadCounters(inputfile,outputfile,alignfiletype=\"sam\"):\n",
    "\t\"\"\" runs pyReadCounters on all the files \"\"\"\n",
    "\toutputfile = re.search(\"(.*)_count_output_reads.gtf\",outputfile).group(1)\n",
    "\tcmd = \"pyReadCounters.py -f '%s' --gtf '%s' -v --rpkm -o '%s' --file_type='%s' \" % (inputfile,args.gtf,outputfile,alignfiletype)\n",
    "\tlogger.info(cmd)\n",
    "\tos.system(cmd)\n",
    "\t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db1df1af",
   "metadata": {},
   "outputs": [],
   "source": [
    "#here the program is asking pyReadCounters to not track any mutations (it can track only deletions/or specific mutations)\n",
    "#the --blocks option states reads with the same start and end coordinates on a chromosome will be counted as one\n",
    "def runPyReadCountersBlocksNoMuts(inputfile,outputfile,alignfiletype=\"sam\"):\n",
    "\t\"\"\" runs pyReadCounters on all the files with options --mutations  nomuts --blocks \"\"\"\n",
    "\toutputfile = re.search(\"(.*)_count_output_cDNAs.gtf\",outputfile).group(1)\n",
    "\tcmd = \"pyReadCounters.py -f '%s' --gtf '%s' -v --rpkm -o '%s' --file_type='%s' --mutations  nomuts --blocks\" % (inputfile,args.gtf,outputfile,alignfiletype)\n",
    "\tlogger.info(cmd)\n",
    "\tos.system(cmd)\n",
    "\t"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0491b746",
   "metadata": {},
   "source": [
    "pyGTFsgr is a tool that produces sgr files from cluster, motif or interval GTF files and these can be loaded into the genome browsers to visualise read distribution over the genome. \n",
    "Inputs:\n",
    "* a GTF file with intervals\n",
    "* tab-delimited file containing information about the length of each chromosome (this is -c in the parser and refered to in the code)\n",
    "The --zeros flag asks for chromsomal positions without any covergae to still be reported\n",
    "By default, the script assumes that each interval is unique and ignores the number of identical reads reported by pyReadCounters. The --count flag asks to include duplicate reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30e9a3d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def makeCoverageSgrFiles(inputfile,outputfiles):\n",
    "\t\"\"\" takes the pyReadCounters gtf output files and generates coverage sgr files \"\"\"\n",
    "\toutputfile = re.search(\"(.*)_plus.sgr\",outputfiles[0]).group(1)\n",
    "\tcmd = \"pyGTF2sgr.py --gtf '%s' --zeros --count -v -o '%s' -c '%s'\" % (inputfile,outputfile,args.chromosome)\n",
    "\tlogger.info(cmd)\n",
    "\tos.system(cmd)\n",
    "\t\t"
   ]
  },
  {
   "attachments": {
    "bedgraphPlot.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAABdwAAAPoCAIAAABLbKuyAAAgAElEQVR4nOzdeZQU1aE/8BoYFAFFQHBB3BAUMTwFl7gk0QQ1LkeMWxQ1xhcEo4C+l3ii5vlc8YlRX9yionFJiGgSk6jRJAqKRNBEwcQloIILICoimygMMDO/P+qlf53pZbp7uvt293w+h8Pp6a7ldnVV3dvfvnWrrrm5OQIAAACgvDqELgAAAABAeySUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAA9aELAAAUoq6uLvurvXr12nXXXQ877LAxY8b07du3bAVLLUn8oLm5OfuUK1asePzxxx999NH58+cvWbLkk08+6d69+1ZbbbX33nsfdNBB3/zmN3v37l368pZK7tsBAGg/6rQMAKAaZQ9lkm2yySbXXnvtBRdckPssRZRLGLF69eorr7zy5ptv3rBhQ6ZpOnXqdMYZZ1x77bVVGs0IZQCAVEIZAKhKiS/5Xbt2bfFSY2PjunXrWjw5YcKESy65pBwl+1ethhEzZ8484YQTPvrooxbPd+zYsbGxscWTPXv2fPjhhw855JBiF7PkhDIAQCqhDABUpexf8leuXPnee+/9/ve/v/7661euXBlFUYcOHV599dU99tijrKVsrZxPPvnkcccdt3bt2vjPfffdd+TIkUcccUS/fv26dOmyePHit95669lnn7355ptXrVoVT9O5c+cnnnji0EMPLU/5i0UoAwCkEsoAQFXK8Uv+Bx98MGTIkGXLlkVRdM4559x+++3lKFySLOV8880399prrziR6dWr1y233HLKKaekvcZq5cqVV1555f/+7//Gf26zzTavvPJKdV3HJJQBAFK5+xIA1LJtt932uuuuix9PmzYtbGGSbdy48YwzzogTme233/6555479dRTM416s+WWW95www0XXnhh/OeHH354xRVXlK+sAACloacMAFSl3HtevPnmm7vttlsURZ07d05cKFQ2mcr5s5/97Mwzz4yiqFOnTi+88MLQoUNbXVRzc/Phhx8+derUKIo222yzRYsW9erVqwRFLgk9ZQCAVHrKAECN22677eIHm222WaZpPv/88zvuuOOYY47ZcccdO3fuvPXWWx9wwAHnn3/+m2++2eryX3rppbFjxw4aNGiLLbbo2rXrwIEDR40a9dxzz2WZpbm5+aabboofX3LJJbkkMlEU1dXVnX/++fHjtWvXPvnkk6kTxOI/586de/bZZw8dOnSnnXb67LPPkqf8+OOPf/SjHx111FFf+MIXttxyy27duu2+++5f+9rXrr766o8//jjT2pMX/txzz40aNWrAgAFdu3bt3r37HnvsMW7cuNmzZ+fyRqIoevXVV8eOHTtw4MDNN9+8R48e++233+jRo+fPn59p+qampoceeuiEE07Ya6+9Nt9882222eaggw4aNWrUrFmzclwjAFCJmgGAKpR7Vf63v/0tnnKvvfZKO8EjjzySCG5SnXnmmWvXrk0747p160aNGpVpxu985zsNDQ1py5kIL7p06fLJJ5/k/q43btw4fvz4MWPGjBkz5qGHHmrxavK6HnzwwU022STxzKeffpqY7NFHH+3Ro0emYm+66aYTJkxoamrKtPB169adddZZmWYfPXp0Q0NDasmTy3b99dd36tQpdd76+vrrr78+dd45c+b827/9W6Y1nn/++amlBQCqQn2mCh4AqA2PP/54/OC4445LffWee+4ZNWpU8z9Tg/r6+t122+3jjz9eunRp/Mz999+/YMGCJ554YvPNN0+ecePGjSNGjPjTn/6UeKZ79+7bbLPNJ598Eo8r/NOf/jT1ztyxZ555Jn5w4okn9uzZM/f30rFjx0QXmyyeeuqpkSNHNjU1RVHUr1+/3Xffvb7+/9o8L7/88vHHH79x48b4zx49evTr16+pqWnRokXxDZ4aGhp++MMf7rLLLqecckrahZ9yyim/+93v6urqTjzxxKOOOmqrrbZ6//33H3vssXg7T5o06f3333/kkUc6duyYdvYbbrjh+9//fhRFXbp0GTZsWM+ePV999dW33347iqKNGzdeeOGFQ4cOTb631JIlS4YPH758+fL4z549e+6www6rV69+77334luG33TTTTvvvHOiDxEAUE1Cp0IAQCFyrMr//Oc/xx1GevXqtWTJkhavvvTSS4kuGzvttNNjjz0W9/KIQ4qzzz47sZYLLrigxbxXX3114tUDDzxwxowZjY2N8bxz5sw57LDDsjQ5RowYET957733tnE7JEusq2fPnptvvvlPfvKT5A4ysaOPPjqepn///k899VSij0ljY+Mf/vCHvn37xq/uv//+mRYeRdHmm2/+7LPPtpjgD3/4Q5cuXeIJbrjhhiyzd+jQ4dprr92wYUP8UlNT0wMPPJDo1zNixIjkGb/73e/Gz/ft2/dPf/pTosALFy488sgj45f69etX6DYDAEIy0C8AVKXkkVNavNTY2Pjhhx++9957v//973/7299GUdStW7fHHnvskEMOaTHloYceOn369CiKhgwZMnPmzG7durWYYPLkyWeccUYURR06dHj11Vf32GOP+PmVK1duv/328UAtxx577K9//esW1+M0NTWdfvrpU6ZMSTyT3OQYNGjQvHnzoih69dVX99xzz0LefzqJbVJfXz9r1qx99923xQRr167t3r37hg0boih67bXXBg8e3GKC6dOnx71UunTp0mIYmuQ7Q02ZMiVtP5rE6MXdu3dfsmRJIqPJZfbrrrvuBz/4QRRF22+//aJFixLP77DDDvGf06ZN++pXv5o8y+rVq7fZZpt48OaPP/54q622Si0SAFDJhDIAUJUy3T06Vbdu3Z5//vnU7OOVV16JRyqpr6+fPXv2kCFD0s5+6qmnPvjgg1EUXXPNNRdffHH85G233TZ27Ngoijp37jx//vxEB5NkS5cu3XnnnT///PP4z+QmR58+feLxdJcuXdq7d+8c30irEttk9OjRd955Z+oE8+fP32effaIo2nPPPdMORbx+/fpNN900tcDJCx80aNBrr73WoUOauyU0NjYOHjz4jTfeiKLo3nvv/fa3v506+wEHHDBz5szUjy9xk6wWq950003Xr18fRdHKlSu7d+/eYq5rrrlm4cKFURRdeOGF/fv3Ty0SAFDJ3H0JAGrcmjVrDj/88MmTJ7d4PjEczJe+9KVMiUwURYlBbeM+NbHEoDBjx45Nm8hEUdSnT59zzz037UsrVqyIH2y55ZaZ1jtv3ry61mSaN+3oOVEU7brrritXrly5cmWmm0OtWbMm0zITxo0blzaRiaKoY8eO48ePjx8nNlHq7GlLvuuuu6adfocddogfxNFYC5dccskdd9xxxx13SGQAoBoJZQCgumW6RHn16tUvvPDClVde2blz5w8++OCMM864/vrrk2f8y1/+Ej/4yle+kmX5cdeSKIoWLFiQeHLOnDnxg+QhaVNlejVxXc+nn36a7b0VKhFk5K65uXnevHlZbiaVkHrRU7JBgwbFD1566aXsE7SQKehJXLL03e9+91vf+tb06dPjy68AgBrg8iUAqEqJ3hatVuVz587dZ5994suIZs+ePXTo0Pj5L3/5y3/+859zX2Pv3r0Tt2TaYost4jzlzTffHDBgQKZZMl2SM2DAgPnz50dR9PrrryfGqWmhoaHhnXfeSX2+ubk5MUumK4yWL1+e5abXsfnz57/44ovz589/66233nrrrXnz5q1cubLFitIufOHChf369cu02IULF+64445RFPXq1Su+C1WL2VevXt3iPlap0ySveuXKlcOGDYtvzxTr1q3bAQcccPDBBx911FHDhg3L/UI2AKDSCGUAoCrlHspEUXTttdfGw8GcfvrpP//5z+Mn99prr7///e+5r7Fz587xmLJRFNXX18f3Y/78888322yzTLN8/vnnXbt2TS3n8OHDp02bFkXR3Xff/Z3vfCf3MkRRtGzZssQwNJlyk6ampkxRRVNT0+TJk//nf/4nHmm4hQEDBrz11lvZF579LX/22WfxeMn19fXJXVpy+bwyTbNmzZqrrrrqtttuazH2cBRFO+2007nnnnvBBRe0GGgZAKgKQhkAqEp5hTKvvvpqPGrMbrvtlggjEj1lrr766hNOOCGXle6+++7xg169ei1fvjxqrafMggULEkOlJJfzmmuu+eEPfxhF0YgRI373u9/lsuqEWbNmHXTQQanLjHLYJo2NjWeccUbyPaF23XXXYcOGDRo0aLfddhsyZMigQYMSlxFlWvgbb7wxcODATMV77bXXvvCFL0RR1L179+SuN20JZWLr1q2bOnXqo48++swzz8T9jBKGDRv27LPPJvIvAKBa1IcuAABQcrvsskv8IL5TTyzR32TVqlWJtCVHvXv3jkOZt956K0so0yI7SDjiiCPiUObRRx+dO3dupmFW0vr1r3+dV1GT3X333XEis8kmm1xyySXnnHPO1ltvne9C3n777SyhTOItF/0G1Z07dz7mmGOOOeaYKIqWLFny9NNP/+pXv3r00UejKJo9e/Zll13WYswgAKDyGegXAGrfokWL4geJ+z1HSSP4vvbaa1nmbWhoWLx48eLFixO3TEqeN9M9hmIzZ85M+/zQoUPj3i7Nzc3nnHNO7iPXvvfee3fccUeOE6eaNGlS/OAnP/nJZZddlprIJK7PymLGjBlZXk3c0yoxdk8bxbeLStxZPLbddtudfvrpjzzyyF133RU/8/jjjxdldQBAOQllAKD2PfXUU/GD5BsnDx8+PH4wbdq0999/P9O8//3f/92vX79+/fold1FJ3BLo1ltvXbx4cdoZly9fftNNN6V9qa6u7vLLL48fz5gxY8yYMbnkMqtWrTr++ONzyU0yefPNN+MH3/jGN9JOMGvWrFYXcsstt3zyySdpX1qwYMHdd98dP85+X6rcDRgwoEePHkceeWTaV0866aT4QdpBkQGACieUAYAat3jx4ssuuyx+fPTRRyee32effeIOL+vXr//2t7+dNux45ZVXbr311iiKOnfufPLJJyeeP/nkk7fYYosoitatW3feeeelRirNzc3f//73V69enalUw4cPP++88+LH995772GHHZZ25N2EF1988Ytf/GLiVtyF2XLLLeMH8bVXLXz44YdjxoxJ/Jlp8Jc1a9Z8//vfT3113bp13/3udzdu3BhFUbdu3UaOHNmWoibEt6+aOXNm2vDrr3/9a/xgm222KcrqAIByEsoAQG1qbGx866237rzzziFDhsRXHnXr1m3s2LGJCerq6q677rp4cNmpU6cOGTLkySefTMQr69atu++++w455JD4wpmzzjqre/fuiXm7dev2X//1X/HjRx999JBDDnnuueeampriZ1577bURI0bce++9yVdLpbrxxhsTPW6effbZwYMHn3jiiffff/+CBQvila5fv/6dd975+c9/ftRRR+23337z5s2rq6v78Y9/vN122xW2Tfbaa6/4wYUXXrhmzZrkbfXQQw99+ctfXrBgQeLJTAPiRFF03333feMb33j99dfjP5uamp555pkvfelLiR5JF198cfLmaovBgwfHJTzmmGOS75a1cePG3/72t6eddlr8ZzzWDABQXdx9CQCqUuJOPZnuudPQ0BD32kj45S9/mbjaJWHixIkXXXRR4s9OnToNHDiwqanpvffeS4xjMnTo0GeffTa+03NCY2PjiBEjkocy2XLLLbfZZpsVK1Z89NFH8TP333//mWeeGT9O2+RoaGg455xz7rvvvtSXOnXq1KIDTpcuXe69996TTz75G9/4RnzPpnzvvjR16tTDDz88fnXzzTf/4he/2KtXryVLlsydO/fjjz+OomjQoEENDQ1vv/12PMGhhx561VVXxTeuSiz8hBNOePjhh+PHW2+9dY8ePT744INVq1Yl1nLEEUf8/ve/r6//l9spFHz3pVdeeWXo0KHxDcijKOrXr1/v3r0bGhoWLlz46aefxk/26dPnlVdeKWDQYgAgsGYAoArlVd1379794YcfzrSoe++9N3FdTwsdO3Y8++yzV65cmXbGhoaGs88+O+2MnTp1uvXWW5PLmWntTU1Nv/vd7/bcc8/sb2HkyJGLFi2KZ7nmmmvSLjOX5s21116buOl1C6eccsqKFSvuvffe5CefeuqpFgtfv3598lVOLfz7v//7unXrsnxeWcqWaZrJkydn6XN0wAEHzJ07N8tiAYCKpacMAFSlRK+KLBP06tVr0KBBRx999JgxYzLFLrEVK1bcc889TzzxxNy5cz/55JPevXsPHDhw8ODB48ePz3LH69icOXPuu+++qVOnvv/++w0NDX379h0+fPjYsWO/8IUvRLn1EImiqLGx8c9//vMTTzwxY8aMJUuWLF26dJNNNtl666132mmnI488csSIEclDFE+bNi0epbg5z54ysXnz5k2cOPHvf//7ggULmpub+/XrN3z48G9961vDhg2LJ7jjjjuuu+66ZcuWDR48OL7+K3Xhs2bNuvvuu6dPn/7BBx/U19dvt912X/va184666x999037UoL7ikTmz9//h133DFnzpz33nvv/fff79mz50477dS/f/9Ro0Z9+ctfbnVnAAAqk1AGAKB1OSY+AAC5M9AvAAAAQABCGQAAAIAAhDIAAAAAAQhlAAAAAAIQygAAAAAEIJQBAAAACEAoAwAAABBAXXNzc+gyAAAAALQ7esoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAtDuzo7rZUV3oUtDeCWUAAAAAAhDKAAAAAARQH7oAAEAVS3T8HhY1hy0JAEDV0VMGAAAAIAChDAAAAEAALl8CANrKtUsAVB2VF5VATxkAAACAAIQyAEBbJYb7BYBqMTuqU38RnFAGAGgrPcABAAoglAEA2sovjQAABTDQLwDQVnrKAFB1VF5UgrrmZjsiAAAAQLnpKQMAFC5x4ZLfGwEA8mVMGQAAANodd1+iEugpAwAUTgcZAICC6SkDAAAAEIBQBgAAACAAly8BAIUz0C8AQMH0lAEAAAAIQCgDALSVbjIAAAWoa27WigIAAAAoNz1lAAAAAAIQygAAAAAEIJQBAAAACEAoAwAAABCAUAYAAAAggPrQBQAAqtjsqC5+4K7YAAD50lMGAAAAIIC65ma/awEAAACUm54yAEDhZkd1iSuYAADIi1AGAAAAIAChDAAAAO2Ozp5UAndfAgAK56ZLAAAF01MGAAAAIAB3XwIAAAAIQE8ZAKBNXJAPAFAYoQwAULg4kZHLAAAUQCgD7ZcB54G2iwf6NdwvAFVHY5hKIJSBdkoNBBSLRAYAoDAG+gUAAAAIQE8ZaKfi7pr6ywBBOPkAAERCGQCgLQqId40NDAAQqw9dACAMY0AAAACEpacMAAAA7Y5r+akEBvqFdipRA+kyAwBAO6Q9TCXQUwYAAIB2J85iJDKEpacMAAAAQAA10lOmLp2ePXvut99+11xzzapVq1pMP2/evHiac845J/sy582bV+KyA0D7El/D7zJ+AIAaCWVinTp12vSfOnbsuGLFihdffPGHP/zh7rvv/vzzz6edZdKkSXPmzClzOQEAAABqKpR55ZVX1v3Thg0bPvroo8mTJw8cOPDDDz88/PDD//GPf6TO0tzcPG7cuKampvKXFgDaJ1fvAwDEaiqUSVZXV9enT5/TTjtt9uzZBx100Jo1a0aPHp06gM622247a9asyZMnBykkhOXaAaAo8j2ZOPkAUAlcS0slqNlQJqFbt24/+9nPOnXqNHPmzL/+9a8tXr3uuuuiKPrBD36wevXqEKWDYOIaSD0EtFEBJxN3uwAgOM1gKkTthzJRFO2yyy7HHHNMFEUPPfRQi5dOO+20gw466MMPP7zyyitbXU5TU9Ovf/3rQw89tG/fvptuuukOO+xw2GGH/epXv2rRAWfp0qUXXXTR3nvv3b179549ew4bNuy6665bs2ZNi6X97W9/O+200+JF7bzzziNHjkzNjKB0fB0CiqKwk4lTEABhqYmoEO0ilImi6KCDDoqiKDX1qKuru+WWW+rq6m666aa5c+dmX8jll19+0kknTZ8+ffXq1dtuu+1HH300derUk08++aqrrkpM88477wwePHjixIl///vfu3Xr1qFDhzlz5vzgBz844ogjNmzYkJjsnnvuGTZs2AMPPPDhhx/27t174cKFU6ZM+eIXv3jLLbcU700DQMnFvzRq2gJQdYZFzeovgmsvocwOO+wQRdGCBQtSX9p7773HjBmzcePG8ePHpw46k7B48eI4fLnrrrtWrVr17rvvLl++fPz48VEUTZw4cePGjfFk48aNW7Zs2WGHHbZo0aL3339/2bJl06dP79at26xZs6ZMmRJP8+KLL44ePTqKomuuuebTTz9dvHjx6tWrJ0yY0KFDh/Hjx8+cObPY7x7S0GMTAIB2yFAyVJT2Esr069cviqIVK1akffXqq6/u2bPn1KlTf/vb32ZawgsvvBBF0Ve+8pVRo0Z16NAhiqKuXbvGMc3nn3/+zjvvxJPNmjUriqLLL7+8b9++8TNf+cpXLrjggmHDhi1btix+ZsKECY2NjZdeeunFF1/cpUuXeFGXXHLJpZdeGkVRLhdSQdsZ0wEoCj8zAgAUrL2EMnV1dYn/U/Xq1evqq6+Ooug//uM/Pv/887TTHHfccWvXrn3qqaeSnxDZtRkAACAASURBVFy3bl38oLGxMX7QrVu3KIoefvjhxEtRFF111VUvvfTSf/7nf0ZR1NzcPGPGjCiKzjnnnBarOOuss6IomjlzZqLfDZSU71FAEPFPlH6lBABoL6HMwoULoyjacsstM00wevTovfbaa+HChRMnTkw7QX19fefOnTt16rRy5coXX3zxgQceOP/88/fdd98Wk40dOzaKohtvvHG77bYbM2bMgw8+uHjx4uQJPvnkkxUrVtTV1R1yyCG7/6vDDjssiqLPPvts1apVbXy/AAAAZOEXAipBewllFi1aFEXRrrvummmCjh07xoPsTpw48e233047zWOPPbb//vv36NFjv/32O+20026++eZevXq1mObCCy+cMmXKnnvuuWLFikmTJp166qn9+vUbOnToXXfd1dTUFEXR+++/H0VRc3PzGynefPPNeCHuzw1ADdNND4CAWlx4K5chrPYSyvzpT3+Komi//fbLMs3BBx982mmnNTQ0xNcZtfDAAw8ce+yxf/3rX7/+9a/ffPPNzz333PLly+fMmdNisrq6ulNOOeXVV1995513Jk2a9O1vf7tPnz4vv/zy6NGj40408eg2Xbt2bWpqas5g5513Lto7B4BSKuBaJM1fAAJKVFvGWKQS1GW531AViQeLmTt37u6775766pw5c4YNGxZF0V/+8pc4l5k3b96gQYOiKGrx9pcsWbLbbrutWbPmj3/849e//vXkZe6xxx5z586dMGHCJZdckpj+s88+iweRSUz27rvvduzYMU5eYhs2bLjzzjvHjRtXV1f36aefdu3atVevXsuXL3/nnXd22mmn5LU3NDQsWrSovr6+xfMAULESCUtejdrZUZ1GMADlV1i1BaVT+z1lPvvss3HjxkVRdPDBB6cOAdPCdtttF98CKb7XdbL4/kqHH3548pN/+MMfWkx2wAEH7LzzzitXrkw806lTp1NPPTWKoubm5ngU4XjsmBtvvLHFvDfddNOAAQOSQx8AqHBxJ/B8m7aawgAEoQKi0tRsKNPc3Lxs2bIpU6YMGzZs1qxZ3bp1u/POOzPdfSnZBRdcMHDgwMTwLgkDBw6Moug3v/lN3Lmmubn5kUceGTVqVPxqPFJMFEVDhgxpbGw877zzErffXrp06QUXXBBF0aBBg3r37h1F0YQJEzp16nTbbbddfvnln332WRRFGzduvP/++y+99NK6urpzzz23KFsAWmVsMwAAgIBq6vKlTp06dejwfzHTxo0bE7ep3nrrrX/zm98ceOCBiekzXb4U++Mf/3jkkUfGjxPXJU2ZMmXkyJFRFPXr12/bbbd99913ly5devLJJ7/++uuvv/56165dx48ff80117z88sv777//hg0bOnXqtO222zY2Nn7wwQdNTU1du3adPn36PvvsEy/2tttui4eYqaur23777ZcvXx6nM7feeut5551Xkm0E/0rXTQAA2qHkAWUguJrqKbNhw4aGf2psbOzevfuwYcOuuuqqefPmJScyrfr6178+YsSIFk+eeuqpjzzyyMEHH/zZZ5+9++67++677y9+8YsHH3xw8uTJQ4YMqaurW79+fRRFe++995w5c0aOHLnjjjsuXbp07dq1Q4YMufDCC+fNm5dIZKIoOu+882bNmnXCCSf07t37o48+2nbbbU8++eSXX35ZIkPZqIcAAGjPdBunEtRITxmgAH4lAACgfdJtnApRUz1lAKqIH2cAAEKRxVAh9JQBCMCPMwAAgJ4yAAHIYqgZBfT5imfRUwyAINRBVBShDEAAmgIAAIBQBiAAPWWoMXnljMOi5vhf6coDAFAVhDIAAAAAAdSHLgAQjFtiB+TyJWqGcwhFpGICoL3RUwbaKaFAWPFXDl88ABJUTEB5JBpgBp6nEghloJ0SCgBQUVRMQNk41VA56pqb7Y4A5Zb4TUabAACg/DTGqBB6ygD50cmzKFT/AKlUMUDZaIxRIQz0CwBAeOIYoDwSY4rLZagEespA1SvzEGVqr6Lw3YP2zMCKpKV+AaAd0lMGyI/vUQCUiFwGgPbGQL9AfgyKBiRLdAIPXRAAyImai4ri8iWgEKqxtnP5BkALTowAtDdCGWi/2tLw1WgGYgZKpFjULEB5qLmoKEIZqHqFDZkZT19wC1hNVhQ2I0BCfEp0YgRKTac8KopQBtoprd7gfATUDE1bisWJESgndwOkEghlAALQCKBmFNbtzs5PWk6MQBm4fImKIpSB9qstFZJGMxAr4DTSxssnqVV2ieKScEEmiaMjbgwLaAhLKANVL0hFovZqI0MnUEsKa9Ha/2nBLlFE4hjIJHF0OEyoEEIZqBGF/VitNgrL9qd98rMkmdg3gFJLnGScbagQQhmoei4EAIrL+QQAoDyEMtB++UEyLNufylSenFfuAyWlfoFWuesCFaI+dAGAKqOdByRLjJWY7yyzozrnEygdxxdAVRDKQNUbFjX7blNEBXzDhNpTukPAb5JQBokDTXUGqRLHhSY0lcDlS1ALylmd6OpZLLYhlals5xPtYFI5MQLllJzOOP8Qip4ylIruBuVU2Nb240BALt+gMiXfKLREO6d9nkycGIvINoRW6VBGhdBThtISOVesNo7lqfYCCqO3HZmoWYrLUQaZqIaoKHrKAPnRaC4Km5HKFI9RFdlFocrpdgSZ6CBDpdFTBtqpgushvy1ADUu+fKmwGQGgksVt4PhHCM1aKoFQhtKSQJeBuqSIyrkxfXDkoobbi7X6viiYXQIoD99QqChCGUpLA6sMEnl/OVdaq59s2TZjG8f0ob0p/66S17EwLGrWwIWKEh+VDkzIwmFChTCmDFS9gq8bVwmlkpJALK9TinErKAp7EZCqFCOdpS7T+YeA9JShJAygVcOCdMwpm7K9r1rdgBRdOX/Ha0soWcC8iev5C14ptccuASQrxQkh7ehpTj4EJJSBqteWIXvLvEaS6TFLLsrZTHQVJNQSXzIppyra3xKVnWYYFUIoQ/FVyxm5lhTw9b7gMU2MVA/lV/lHnKYtVJTKP2lALkrUQTuxQG1aKoExZSi+uEd6pI1eLm25WMxnFIpL/ABSOSVSMI3PmlTwyIk5Ljb5TwhFKENJqBGDyKs5khjNwfDAULHKHN5plULNUFlD7hK/KEMQLl+iJHQFLIPERs6xY2exPpTa/nBrdfAOyEUB96EobHzf5Me5rChx1WSL84+rKQEqjdMy5EsoQ/E5EZdZYT0w2/gx1eqnXKyspNXtU6sbkHIqesO3Rd5RxCVnWUthI1u1+gzVSMpW9CO6PW9Myql0e1ruDbO8TiB+G6tA7fmUJZSh+AydRTuXZRDlmjwoavJNVZosWzjsxteuhWIpeAB+qGElOiIcaAGljc/a+SdiTBmoerlcB9vOz3QVyFdZ8lL5O0zRR2GkHbILFZftSfkld9yu2D0wbXfLii1tO9HOh/XRU4bScoIrv+zbvD2f73JRtu2jd0kmtky+2sNptj28Rygbp1nKoFjn7XjosVLXArmPblbSYtBuCWWg1qStMJIrm6LUKLX6HSnH8ZIL24YGwmiVDZJJxV6+lGnteR0mrU5sx2gnjCkTFXVos7Qbs51vXsqmuva0TMPJQ3kIZSitKj2vVdcZuSjDZJZ6xupSosipSpOs7IdDSXeJqtjf8j1dJHftzjLkbfnPQoWtroh7db6bsSp2D8hXUXbsaj86cil/Ygiean+zpVO6iKGcm71Y7yLfnwpaLU9bQpxq32+ruvCVSSjTXhR8yihFYQqQ71mvKD+6Vs7bz6Lt7yXLeLQFfxXPsTlVrFq2jUtIXVouoy3mXvjct1XAXa7grgol+k6eKE+LCKNi+1PktTOkfujJb7OwtQd576lBUuoEUfEO9rwKU9IVVdRyyix5Owf5VTnTwVLtX3KKqNq3Q6YquMW+lzpNwF8+qn2bt0XB1VbB87alGAWfKApYfquzVOluk/aza/sZuEq3RrEIZWpW2oZyAV9vKjPKSf3yltr0z5I1pH21pBlBqwFHW+L2LAvMUqTs8yb/mfb57EvId6W5FKnghCj7YjMtJ8d0Jq8aN9PEadeVNqFIW+zsay/6dstx4QUsrY0rKmzHS3vqyOUrQaZFZXk1087Qxs2YfMO7tEXK/YSf/aVc9sM2Hv5FPA1mX0vaP1s9J2SvXwooQ+W3QbN/0AXM2MYypD1PZjpbVpoSFSzLTpv8ZAG7XJZPP8elpZ738j3V51XytOfzFs9kmiuXJWcqZ47Vd6YlJ//LsWCZzvaZ1lvc4zF1sQXXvJnmzbJNcnkvmWYvWI7LSa1NinLE5asoH32L95L9Y8q+nOxbozz1fuWra26uyu70pJW6Q7c6kHWL1nxi+uQZc/zlIZfDqcWikmfJ8r2iLbIsNvdRvpM3S44lTLslU0uVfWlFGQgmr8HM2z7yeY5LKMoQ63l9glHSTh4lbc/cP9O81ljcedMuJHXfTj1UWxxiBRSgxVyFbfMsT7blwM+lMFkOwLTzZlpmjvtJXrtTloVkWkIR701QlKKWVPazaPYZ4wfZT/t5nc+zTJnvlsz+dtJWuGkPmbwmiP71BJjvMtso02Zv9ZTVlmKkrbgLPrhynDG11s60wVO3QJamQomO01Z3hrz26uzTt/HU2uqiilXDRll3wrQT5Nv6zfe0llctXKyjpihLK2Dt8YN811iiu/YUvcJNyLEtkaqNO09e8xY2fdtnTMyeeJypvqgxQpmaUrrzZhnOy+38RmiUmh0MqBkFx6OVH8ZVtXxT4AqkroTKUfTbdFSvmg9lXL4EQDYVWxFWbMGqnQ0LAFA2esoAAAAABKCnDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyALTNpLpoUl3oQgAAQPURylAyvqcBAABAZkIZSkwuAwAAAOkIZQAAAAACEMoAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAAAAAAEIZQAAgCo3qS6aVBe6EAB5E8oAAAAABCCUAQAAAAhAKAMAAAAQgFAGAAAAIAChDAAAAEAAQhkAAACAAIQyAP/KPTUBAICyEMoAAADQjvlNjnCEMgBJ1McAAEC5CGUAAAAAAhDKAAAAAAQglAFIMro5dAkAAID2QigDkMSYMgAAQLkIZQAAAAACEMoAAAAABCCUAQAAAAhAKAOQxEC/AABAuQhlAAAAAAIQygAkcfclAACgXIQyALSBGAsAAAollAFIYkyZfNliAJVJaA5QDYQyAEk0YQGoAXF1plIDqHhCGYAk+n0AAADlIpQBSOJHxXzZYgAAUCihDEASPWUAAIByqQ9dAIBKot8HUAaJU40gGADaNz1lAGgDXykBAKBQQhkAAACAAIQylIZrQKhS+n3ky8FOEJPq7HsAQA0QylACGspUL3svVD7HKQBQK4QyALSBvkUB6S0C5Cs+bzh1AFQMoQwl4EsatB9a9sH5CAAAqpZQhhLwDQHaDyEsAAAUSigDAAAAEIBQBqAgrsmP2QgAAFAooQwAAABAAIWEMtOnT//Od74zcODArl279u7d+8ADDxw3btzixYuTpzn44IPrcrD99tsXsPBYqws/5ZRTCnh3eRVj3rx58boKXhFQ3fQTgfIzkhEAUCvq85p6zZo1o0aNeuihh+I/6+rqPv/882XLlj3//PM//elPf/SjH5133nnxS5tsssmmm26aPG9DQ0MURZ06derQ4f8nQcnT5L7wZC0W2OKlvN5dW4oB8P8lkhpfHaGk4mPNgQYAVK08eso0NDQcffTRDz300BZbbHHddde9++67GzZsWLZs2bRp04YPH7527dpx48ZNnTo1nvjpp59e96/i51955ZXkJxcsWFDAwpO1WGCyn//85/lujoKLkZ0ONQBQNC16qOmwBgBUrTxCmQkTJsyYMWPrrbd+6aWXLrzwwh133LFjx469evX66le/+uSTTx5//PHNzc1nnnlmY2NjAeUo6cKrrhhA1fATPcHZCQEAqlauocyyZcsmTpwYRdENN9wwYMCAFq/W1dXdfvvtHTt2XLJkyXPPPZdvIUq68EooxlZbbXXZZZdddtllRSsrAMT0EwEAqFq5hjK/+MUv1q9fv8suu3zzm99MO0GfPn3GjBlz0EEH/eMf/8i3ECVdeNmKMWXKlH333bdLly59+vT56le/+sc//jHx0rJly6644oorrrgi/jO+lGnevHlPP/30Mcccs9VWW2266ab9+/cfO3Zs2iGNq5ufcKltvg8DANQG7TpCyHWg36effjqKotNPP72+PuMst912W2GFKOnCy1OMq6+++tJLL+3Ro8egQYPmzp37zDPPPPPMM7/85S9POumkTIu68cYb77rrrl122eXYY49dv379jBkzbrvttsmTJz/xxBMHHnhg299OpZhUJ5ehmoxuVh8DQJVRdwNVK9eeMn/729+iKOrfv38pClHShZenGJdffvlNN920bNmy2bNnv/POO0OHDo2iaMKECVlmueuuu0aNGjVv3rx77rln8uTJb7zxxplnnrlq1aqzzjprw4YNhb0FoK206gCoduoygOqRx5gyURTttNNOpShEWxZ+5JFH7p7Oj3/843IW48wzzxw/fnx8c+6tt946Hj7m9ddfzzJL//79b7/99sR9uzfbbLO448ybb77505/+tIAyAAAAAFUk18uX1q9fH0XRZpttVopCtGXh7777btrn44SlbMUYPXp08p+77rprFEUbN27MMsu5557b4jqpTp06nXfeed/73veeeOKJc845p4BiAOXm6jwACM7VxxSFdh0h5NpTZsstt4yi6O233y5FIdqy8Llz5zanc/XVV5ezGAMHDkz+M+4yk93gwYNTn9xjjz2iKHrrrbcKKANQBPlWxpqAmi8AVA71Mm1h/yGEXEOZ+BbRCxYsyDLNtGnTjjvuuNNPP725Ob82ekkXXp5i9OjRI9/V9enTJ9OT7733Xr5LA4oj38pYJKH5AkDlUC8D1SbXUOZLX/pSFEXTpk3LMs2UKVMeeeSR5cuX19Xl10Yv6cIrthhp734dX4217bbbtnHhQJmIJKASOBKBmLMBUG1yDWW+9a1vRVH09NNPz5w5M+0Ea9asefzxx6MoOvTQQ/MtREkXXrHFeOmll1KfjFe9++67t335AADQvugpA1SbXEOZwYMHn3TSSVEUjRw5MnXUlebm5osvvvjDDz+sr68//vjj8y1ESRdescW45ZZbVqxYkfzMokWLbr/99iiKhg8f3vblA+Wg8QcAwekgA1StXEOZKIpuu+22XXbZZeHChcOGDZs4ceKCBQs2bty4YsWK559//thjj7311lujKLrooov69+9fQDlKuvDKLMaKFSu+9KUvvfTSS42NjQ0NDU899dSBBx64du3a/v37n3vuuW1ffki+ptJ+aAUCQCiT6v6vItb4pCjsSISQ6y2xoyjq3bv3jBkzTjrppOeff/6iiy666KKL6urqkse7HT9+/BVXXFFYOUq68MosxqhRo+6+++599923a9euGzZsiG/IvcUWW9x///2bbrppUVYRjK+pAACUjcYnRTGpTi5D+eURykRR1Ldv3+eee+43v/nNgw8++OKLL3700Uc9evQYMGDAHnvscdZZZ+2///5tKUpJF16Bxfje9743cuTIH/3oRy+88ML69et33HHHo48++qKLLurXr1+xVgEANUVzGcjOWYK0ErGd3YPKk18oE0VRhw4dTjzxxBNPPDHfGXO5lXVeCy/dvbFzL8buu++ethgtns802aGHHlrSoYuBklO1Q9nETWq/hwMANSSPMWUAaMn3QwAAoFB595SpLjfccMNdd92Vy5R9+/adNm1aqcsDAAAAEKvxUObjjz9+4403cply3bp1pS4MAABQQi4rBqpNjV++dO211zbn5t133w1d2BqiOgQAAIDW1HhPmcpUuiGKK4VRNgAAKD93XwKqTY33lAEAANoLiQxQbYQyAAAAVLlJdQV22NfNn6CEMpSA3yigHdKgAaCc0tY7KiOg2ghlKAHVIdSSgn93AoAy89Ngu6WtQtUSygCQg1bbOtrBAATnmzlQbYQyAMA/6RgFVAs/BlAs9iWCEsoAAFEU+YUZAKDchDIAFIPv8wBANdKGISihDADFoOsvAOXkizRQE4QyAABAZkabAigZoQwAAJCBOAaglIQyAG3gmh0AAKBQQhmANvD7IQBAVfMbG0EJZQAoBvkUAFCNtGEISigDQDH4lQkAAPIklAGgGPzKBAAAeRLKAAAANUG3TaDaCGUAKAbtYACC020TqDb1oQsAmSWqVV/2AAAAqDlCGQCoSHEwLZWmYvntBADazOVLAOno/5wvW6y4bE8AgHZAKAMAVDwpFQBQi4QyAOnojZ+Fr8cAAFAMQhmAdOQOmcRbxvYpp0l1NjgAQE0SygAAAAAEIJQBSMflSwBUndroWFcDbyFWGx8HUGJCGYB0tKIAoPwKuEhWlQ1UM6EMQAZ+4AIAAEpJKAPQvsmeANrCKTQ4VxwD1UwoA9CO+S5RRXxYUIHckK4S2P5ANRPKQFnojAAAAMC/EsoA5E/ElkrvcQCA9kwLuSBCGaAN9AAiwZ4AAO2KdiCp7BL5E8pAjYqryfKcFp18oT3TKAcAKJRQBiB/LtWpDeXMLmuezQjQrjjt04JdolD1oQsA7YAzVNHFmzRgMuIzpWzsbABUoNHNLWuo4M2zMkh+y7X9TsujPewzOdBThqLymzPtRPDKw4EGAJUsUVPXapVdq++LsrEL/ZNQhuJxXFFmAXe5sHu7Yw2ASuOXOWJ2A8iTUAaoWgG7qwTvKRNp9LQblbCzAeRI3QSQJ2PKAFVrUl2w76sanZSNnQ2gQhj/opZUe/WaKL8dsvrpKQM1xK1koP1wpNcGnyMVSFuCmlfAHu64yJ0NlSehDJRR2c5QToWQFy0taG8c9a2yfVrId4PovxBKKXZdhwOlJJShIlXCia/V5lqZu6Wkrivftbd9CW2Zq/2wcdqo3Y7fXFFsipKyeTMpf/3is6B07F1ZtNrGzmWywlZXovNMe2seF/d7UHvbeimMKVOLchxoo9XLYkNdqRivt8XZM7kMyX9mebNtue4337AjVoYNlfp+42fyGl2lsLNeq9Vnvm8/y3bL/tmlRkvl3EVTt0NhBchrt8l9FS02XdVd/Z58+Je02EXcbVrsEnktNtM7LXXTJPctnNdnkVftE8t0bi9MOc8GFX5w5bsp2vh2Uk9oBSww+WtY5Rz+Zf6gA+5XZd5n2i7LebLg3a+Nh0A8e2rBCqvB852mRfM4de2JebMvJO1LqU3x1FfTfiLJJ4cs7bfUDdgil0lt6+ZbkeUubWM7y5LTVmQ5fs8qSoMte1uixW7Z6ioK2z1yWXLyxNlXERVpy1SPuubmGn+H7U6mXblVmU6mLZYWZTgvZ589r3UVJnHGyfeN57uKIi48U+Ud5f/xtbqi7MtpUdWl1uWZCtlq8Ur3cWRfaer35Fx271a3UqvTtEXyPpzvWlotW/aaL3my7A2v1IlbXWMbt1iW5mOLpk+LWZL/b8vak1dX8BKytEpTp2/jBK2Wpy1vJFbOw7kAaYuay/k2WQFVZ/Lhk/snHv3rTp4psYrSnZlTn8y0ruyniFJUynmVIe1JO5bjBsm+tAK0usVSi5HLd+nElLlUN2nnTS1hG+umtF/Yct+2ua+9xTk5x48p3wo90+xZDq5M+17qS2mXH+V5vKedMbWoqQUOcuYvotKVIfupr41VXqjt1paKKcclt+WoL/9mEcpQBQqIRfJVIQckAABAzQueDVWOWg9ljClTQ0p6uKYu3NkBAAAA2kAoU0MqLUGstPIQ5fCh+NSifDZC2zdXbWzw2ngXlen/tXfvUVKUd/6A3x4GuXnjoigKiyJeE8GoiaJEDGLIeqLGNehxlQgSomskGu/Zze4JJl4SjXGJt2RdjcagMetPo4m7Kt6ioqsx3gLeIaCIKyoojAjD9O+Pip12prunp6d73r48z/F4muqqt97q+tbb1Z+prp6Z/tt/PbnGql1Lx1ejJ18ceoC9WS6Nc2g0zpZWp1p88bPfWzP1U8yGdGlmyqUBXm1fXwIAAACIwJUyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiaI7dAQAAoEekUn99kE5H7QcAf+VKGQAAAIAIhDIAAAAAEQhlAAAAACIQygAAAABEIJQBAAAAiEAoAwAAABCBUAYAAAAgAqEMAAAAQARCGQAAAIAIhDIAANAY0um//R+AKiCUAQCAxpBK/e3/AFQBoQwAAABABEIZAABoDL64BFBlmmN3AAAA6ClyGYBq4koZAAAAgAiEMgAA0DDc5RegmghlAACgMfj1JYAqI5QBAAAAiEAoAwAAABCBX18CAIDG4KeXAKqMK2UAAAAAIhDKAAAAAEQglAEAgIaRSvn1JYDqIZQBAIDGII6BnISVxCOUAQCAxpDc6NftfgGqhlAGAAAaQ3ItgCsCAKqGn8QGAACggbl8jHhcKQMAAEADc08Z4hHKAABAY3A5AECVEcoAAEBjcC0AQJURygAAAABEIJQBAIDG4OtLAFXGry8BAEDDkMsAVBNXygAAAABE4EoZAABoDJkb/bpeBqA6uFIGAAAAIAKhDAAAAEAEQhkAAACACIQyAABAjUul/nbHHIDaIZQBAAAAiEAoAwAAABCBn8QGAIDG4JewAaqMK2UAAAAAInClDAAANIbMrXBdMgNQHVwpAwAAABCBUAYAABpDcoGMy2QAqoZQBgAAGkPy9aXMl5gAiK2cocwDDzxwwgkn7LjjjgMGDNhiiy3GjRt3yimnvP7669nz7L///qkibLvttiU0nui08aOPPrqrm1ZMn1evjqcm4AAAIABJREFUXt3VZgEAAICGVZ4b/a5evXrGjBk333xz8s9UKtXS0rJixYr58+dfc801P/rRj04++eTkqY022qhPnz7Zy3700UchhN69ezc1/S0hyp6n+MaztWuw3VOlbWaBNgEAoNql0y6TAagqqXS3v1P60UcfHXzwwQ899NCmm276L//yL1OmTNl2221Xrlz5zDPPXHDBBffee28qlbr77rsPOuig3D1IpUIICxcu3HnnncvSeOEGS1OJNgEAoEfV8a8vJZtWf9tFz1A/xFOGUOZf//VfzzvvvKFDh/7hD38YPXp09lPpdPrII4+89dZbhw0btmTJkl69euXoQcG8o4TGhTIAAJBbKlWfnzx9qAZqU3e/jLNixYqLLroohHDJJZe0C01CCKlU6sorr+zVq9eyZcsefvjhqmocAAAaixv9Qk6plOOCWLobytx4443r1q3bfvvtjzrqqJwzbLnllt/4xjf222+/BQsWVFXjFbJs2bKBAwemUqkHH3wwe/qvfvWrVCq1ww47rFmzJpny6KOPTp06dcyYMQMGDBgzZszUqVMfeeSRGF0GAAAAIujujX7vu+++EMKxxx7b3Jy3qcsvv7wKG6+QYcOGXXrppdOmTfv617/+zDPP9OvXL4SwYsWKb33rW6lU6tprrx0wYEAI4Yorrsjcn3jrrbd+9tlnn3322RtuuOGyyy6bNWtWzA0AAABoKL74RjzdvVLm6aefDiGMGjWqHJ3p0cYr52tf+9rkyZNffvnl2bNnJ1O+/e1vr1ix4tRTTx0/fnwIYfHixaeddloIYcaMGe++++6yZcvefffdGTNmhBBOO+20J598MmLnAQAAGouvLxFPd2/0O2DAgJaWlgcffPDzn/98iT3Ifw/d0hpPGhw5cmS7395OnHjiiaeeemppnczXZghh4sSJ2dfsLF26dLfddmtpaXnyySf/7//+74tf/OLo0aOffvrp/v37hxCOP/74X/ziFxMnTrznnntSHx/86XR60qRJ8+bNO/roo+fOndvVHgIAQIOq41+Vome4UTTxdPfrS+vWrQshJF/SKbvuNL548eKc01esWFFyf/K1GUJolygNHz78kksumTlz5vTp01euXJlKpa677rokkQkh3H///SGE0047LZUVx6ZSqVmzZs2bN++xxx4ruYcAAABAreju15c233zzEMJrr71Wjs6Us/GFCxemc/n+979fcn/ytZlOp2+77bZ2M8+YMWPixIl/+tOfFi1adMYZZ4wbNy6Zvnbt2qVLl4YQPvWpT7VbJJmyZMmSJI0CAIAS5fw6RjKx/r6m4QIHoGZ1N5RJfqn61VdfLTDPvHnzDj/88GOPPbarX5WqaOOVlkqlZs6cmTw+8sgjM9OXL1+edHXo0KHtFtl6661DCG1tbW+++WZPdRMAgLpTf7FLYY22vUAd6W4ok9y5dt68eQXmmTt37u233/7uu++mujhcVrTxSlu9evU555yTPP7mN7/Z2tqaPN5qq62Srr711lvtFkmmpFKpjnkNAADwCZmrfqrsr7MAxetuKDN16tQQwn333ffII4/knGH16tW/+93vQggHHnhgVTVeaWefffaiRYtOPvnksWPHPvHEE5dddlkyvW/fvttss00I4c9//nO7RZ5//vkQwvDhw/v27dvDvQUAgFpVZX+dBShed0OZ3Xbb7atf/WoI4Zhjjul485d0On3uuecuX768ubn5iCOOqKrGK+q+++674oorRowYceGFF/785z9vamr67ne/m/ke1oQJE0IIP/nJT7K/cpVOp3/84x+HEA444IAYXQYAoN7V6xUl9bpdQAPobigTQrj88su33377JUuW7LnnnhdddNGrr77a2tr63nvvzZ8//9BDD/3pT38aQjjnnHNGjRpVbY1XyAcffDB9+vQQwpVXXrnxxhvvtdde3/rWtz788MOZM2cmKczs2bN79+59zz33nHTSSStXrgwhvPfeezNmzLj//vs32mij2bNnR94AAADqVTotwoD2HBfEkyrL/XHfeOONr371q/Pnz/9ro6lPNDtr1qxLL720qSl3AJTcYGXhwoXtflW65MY7bbAESZu9e/fOtxUhhPPOO+/MM88MIZx44olXX331Mcccc+ONNyZPrV69erfddluyZMl//Md/nHDCCSGEOXPmzJo1K3l26623ztzZ9/LLL/+nf/qncnUbAIBGlPk6T31/zszcUKZBtheoR+UJZUIIbW1tt95660033fTEE0+89dZbAwcOHD169K677jpt2rTPfe5zhXpQRIbSpcYrF8oUdsEFF5xzzjn33nvvpEmTBg0atHDhwi233DLz7O9///tDDjlks802W7BgwbBhw0IIDz/88NVXX/3MM8+8+uqro0aNGjt27EknnbTvvvuWq88AADSoAiFFPd0ZVygD1L6yhTIAAEBVyBdS1Fl4IZShXOoprKTWlOGeMgAAQA2o3GfO5Nep/QoSQBc1x+5AHJdccsnPf/7zYubcZptt5s2bV+n+AABAD3E5AEDVaNBQ5u23337xxReLmXPt2rWV7gwAAPSE5EqWVEouA1Al3FOGivHlXgCAKHr+njJRTvyy7wPiniB0h/ohHveUAQCAxuAzJ0CVEcpQGW7zBgBQber1DK1etwtoAEIZAAAAgAiEMgAAAAARCGUAAIBycM8agC4SygAAQH2JFY7EureLMAioWUIZgApwx0EAIvI2BFAjhDIA5ZacCjshBiAWV44A1AihDAAA1JdG+8NAo20vUEeEMgAAQC1zZRBQs4QyAFCz/HEYIBgMgRomlAGA2uTuRQAANU4oA0D3pFJyAQBC8DUigC4TygAAtUD8B91X6dCk+wepIx1oMEIZKswfTAAqqkGGWR/SoCxq5VCqlX4CdJtQBgBqmY8uAA0STwP1SChDhfm0AABARTnhBGqWUAYAAKhlrpShMPcqoooJZQCglvkoAgDFkMtQlYQyAFDLnGL6+ydgEABqllAGAADqiIQCoHYIZagw19UDUDk+fEKX1NOJmcMfqAtCGSrM+yUAQJWopxOzegqYgAYmlAEAqp5PX0ABhgigZgllAICqV09/3gc61dWQxRAB1CyhDADUMn8fBgCoWUIZAKhlDfL34XzZk0wKCIYCoIYJZaB+pVKN8mkNqHv5RjOjHFSJJBYpVzhSzKHt8AfqglAG6pQzFWgQ/j5ci4Tm1KWIw5EDCqhZQhkAqG6FP8D7KFK77Lv6YD8m4kaN4mmgZgllAABi8DGyDiQxRN3nMsUHLnX/UgCUm1AGoCS+fUCVKO8He4Xdk7zU9LwSRoyaKNSa6CTRicKpSkIZqljy2cC7LD1M1VFbyliuKh/qnsOcRqb+qUpCGSrAeFcN/CmgNKoXgDrm9ACgyghlqADv99VAuAAADajwCYDTAyjAAUIMQhkACvJ9rsYhUu9hdfCCGx+qjd0BleYoo9yEMlRA9lBVB2ecNBQVm49TkEZgL/ewWn/Ba73/AF3VID+4Rs8SylBhxiwqoXJ/m1Wx+YiriKtdBbpAA+qbNx2iKLLw1CdlJZQByOJdNh8fgIlLBTYCWRtQE4xUlJVQBiCLd9mGknwCtNNrjl1WRg4BqlMxZenvKEBdEMoA0PCkM7Ul5ycxu6/W2YN0lZoB6oJQBsoqc/evMt4GrIwfF8vbMRI+zNMzlBk1p0vDY41e9VCj3S5Ssvuqfxuzy6yiQ2XxjUcZsQsccdVwrlINfShGTXQylsIvTq3s4uojlKEKZEKHuIdx9/vQLu/oUlNdXXWnMxc4hWqXHHVV9D1VpPJ2Mt9WV9VLUSu7hi6pg32aHQd3ujmdzpPv2er/3FgWhT900fPq+GXv5tlCd1aaeVzCqksbCoo/CazaXKZcnY91Tl72M+Gurr0SzVahMn7SKWPLDU8o00i6dFpccvvdbLlKjucezomKSViyTzI6jokdd26R7y6FN7PjU1WygzqV820j5+YU2KKOT5W3JDq++D1/JlQNeWjtyn7pyju6dtpaCSev2V3t/seejo2Xt82O7XQcAzvOXIWK6Wp5aybzz5Jj99KW6tJYmr1UCWssS4EV+Y5Q2qb1vOhpXfFHZb6udmm3lmt4zHlC1R0dB8Psid05MHOWa/bEdoNkzmcz82Sf6FbDaUC+s9bCG57vgO3q5pQ2RnV6gp3dSA8r3Odu9iepoo7blV1jJZyiVPqsu7o1x+4AlZfv5C95kE7nGLJzHmPd70OyrnaraNd4vnVlWsj5zwITM0+V5Y+o2d3OXl2Xzti61JOOo2dm8XavRoFuJJvf6ZtH9rZkz59vZ5WgwD4qcsF25zqFm8r5Ka5jEXacJ3zylS+ywyXMXJpiDofiF+/O2kv7NNXpqrvZw45NhU8WT+HryLq6B4s8AHMuXuDeKF09ocnIt65OW+h4vHR1F3T/5Kn4z3WlrbHjdhU/vhU5Z8cDpPCCmXk6vbyxmE4Wo8AbaOZxydcadGy5qmZLdLoHO7695ltdp++t3Zdv75e23o47Ot8pX8633ewFC3em8Nlm4WXz6bSpnM12OiR22rEC9dCuY4Xb6bScsqcXfzabvR8Lv87FHNcFBuFOzzo6jn7FD2sl7NZOz+Lanat32mZGxzOcYk6kc/ah8Lo6llbhPnfpbajdoNelMbDwbO0GyczL1e516/jJgvxSaS9QPSk8WJR86tADHzLbra47TRX+rFjM+1a+Bst+7tXp23yRLZSxV4W3seOYnu/dJVuBN+mcb4SF33Fz9ipnUwUa6ebrlq/NIkur49rbfUQs8Nk45wtefCUX85bc6afWrr5uOV+ZdDr3lhZoJGf3Oj2PKcvBlbPxTlsusKMLV36nzRZ/XHRcXZfGsXb7rtOPHPn60LFuK/c5tpiPlPl6kvPQK7Iyi1mwjBvepVIs3EJXd1/2s8UPvJ12uN34EPIc7PnWW0Jx5qztLr2YJbwXFKixfEdZWd7CCrxulTvjKmZFxQ9olRs0Sl5jV3tVzIvQ81taWMcDM5Sp9jqdIee7cAntl0t3ariEs7jiQ5mcI0mnnSn5lDUzvZii7ereyQyDxQ/y9UIoUxc6ZpPlVUyRVM9bSLW9n3Wq5jpcQIGBvnIl2ulbWtlXWomz2Eqc/3WptZwxR6iaQ7u0copycFXJEZ0znOr+Pu1qDpi99irU6VHQpaZyLlgl9VBexQQ63fkEXnjVxcdk+eLIbkZaZckRCnyCKleyXFVq8UCoxT5XWlWdGGRU557q/nldWd6y6+YsqN4jC/eUofpU21FXif60a7PaNrlImW6n05/4ZzGLFDNnlxos+TUsvvNF9qRC82cv1XGru/865FtFvcq3gUVueCV2d5eOjq4q8oDq5roKz9xpUz0w2HZ8qkvHTruXsUAVlbyP6uDQ62YZFJiz3Ytf2rhXTPc6DoNlHGMLN1La2FKhgatcev4dtkJ65o2+vC1UbkU5FylmhKyETjvT1We72Yfuv+kXHgDLqFxjWqfvX1VyFFcrV8oAAAAAROBKGQAAAIAIhDIAAAAAEQhlAAAAACIQygAAAABEIJQBAAAAiEAoAwAAABCBUAYAAAAgAqEMAAAAQARCGQAAAIAIhDIAAAAAEQhlAAAAACIQygAAAABEIJQBAAAAiEAoAwAAABCBUAYAAAAgAqEMAAAAQARCGQAAAIAIhDIAAAAAEQhlAAAAACIQygAAAABEIJQBAAAAiEAoAwAAABCBUAYAAAAgAqEMAAAAQARCGQAAAIAIhDIAAAAAEQhlAAAAACIQygAAAABEIJQBAAAAiEAoAwAAABCBUAYAAAAgAqEMAAAAQARCGQAAAIAIhDIAAAAAEQhlAAAAACIQygAAAABE0By7A5TB+++/v2HDhti9AAAAgLLp3bv3xhtvHLsXlZVKp9Ox+0B3nXrqqcuWLYvdCwAAACibz3zmM+ecc07sXlSWUAYAAAAgAveUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQQXPsDhSSSqWSB+l0Om5PAAAAAMrLlTKfsGDBglQqlQmDemxZAAAAoNEIZT7hl7/8ZZRlAQAAgEaTquZvBvXw15fefvvt3Xffffny5SWssTvLAgAAAA3IlTJ/tWLFiilTpiSpSk8uCwAAADSmqr7Rb6WtWrVqwYIFCxYsePjhh+fOnfvRRx/1zLIAAAAADf31paOPPvrmm2/O+VSna+zOsgAAAAA9+vWll19++Tvf+c4ee+wxePDgAQMG7L777v/wD/9w2WWXrVq1qpjFn3vuuW9+85s77rjjJptsMnDgwM9+9rMzZ8585ZVXcs6c+ljyz4ULF37961//zGc+M3LkyDVr1pRtkwAAAABK0kNXyrS1tf3bv/3bRRddtH79+o7PDh069IYbbpg0aVL7zmVdKXPJJZece+65HRdvbm6+8MILTz/99ALL3nzzzVOnTl23bl0y5YMPPth4441DCIsXL165cmX2UnvssUdmqcJb1J1lAQAAAHoilEmn0yeccMK1116b/LOpqWnHHXfcaKONXnnllZaWlmRinz597r///n333fcTnfs4WLn44ovPOOOMEEL//v333HPPQYMGPffcc6+99lpmtnnz5h144IE5l7377rsnT57c1tYWQhg+fPjOO+/829/+tm/fvjm72p0vTPXwb0UBAAAANa0nQpnrrrtu2rRpyeNTTz31O9/5zhZbbBFCaGtru+WWW0477bQ333wzhLDbbrs988wzvXr1+lvnPo45QghNTU3nn3/+6aef3tzcHEJIp9M33XTT8ccfn1z/cthhh912222f2LCPlx00aND69esvuuii4447LrlApgChDAAAANAzKh7KrFu3bsSIEW+99VYIYfbs2d/97nfbzfDkk0/ut99+SbZy//33T5gw4W+dywpl5s6de/TRR7db9oc//OHZZ58dQth2222XLl2a/VRm2ebm5kcffXTvvfcuprdCGQAAAKBnVPxGv7feemuSyAwdOjT5ClI7e+211xFHHJE8vvvuu3M2su+++x511FEdpx9++OHJg9dffz1fB6ZPn15kIgMAAADQY5orvYJ77rkneTBr1qx+/frlnOcb3/jGgAEDQggjR47MOcMpp5ySfdVMxg477NBpBzLBDQAAAED1qHgo8/jjjycP9tlnn3zzTJgwIftbSx3tsssuOac3NXV+pc+IESM6nQcAAACgh1X860vLly9PHnQnHBk1alTJyw4bNqzkZQEAAAAqpOI3+m1ubt6wYUMIoaWlJd/Xl/Ip5ta5+ebJTG9ra8v51aeS11iJZQEAAIBGU/ErZfr37588ePfddyu9rpyKT2QAAAAAekzFQ5nBgwcnDxYtWlTpdQEAAADUioqHMnvssUfy4Omnn843zwsvvHDQQQcddNBBU6dOrXR/AAAAAKpBxUOZzM8q/fjHP16/fn3OeW655ZZ58+bNmzfvnXfeqXR/AAAAAKpBxUOZY489Nrm/76JFi37yk590nGH58uVXXHFF8viwww6rdH8AAAAAqkHFQ5lBgwadffbZyeOzzjrrzDPP/Mtf/pL5faKHH374S1/6UvKz2cOGDTvmmGMq3R8AAACAalDxn8QOIbS2tv793//9Pffck5kyePDgESNGLF26dMWKFcmU5ubm3//+95MmTfpE58rxk9hd2kA/iQ0AAAD0jJ4IZUIIH3300cknn3zNNdfkfHbo0KHXX3/9wQcf3G66UAYAAACoVz0UyiSeeuqp6667bt68ea+//npLS8vgwYPHjBnz5S9/+fjjj994441zdE4oAwAAANSpHg1lAAAAAEhU/Ea/AAAAAHQklAEAAACIQCgDAAAAEIFQBgAAACACoQwAAABABEIZAAAAgAiEMgAAAAARCGUAAAAAIhDKAAAAAEQglAEAAACIQCgDAAAAEIFQBgAAACACoQwAAABABEIZAAAAgAiEMgAAAAARCGUAAAAAIhDKAAAAAEQglAEAAACIQCgDAAAAEIFQBgAAACACoQwAAABABEIZAAAAgAiEMgAAAAARCGUAAAAAIhDKAAAAAEQglAEAAACIQCgDAAAAEIFQBgAAACACoQwAAABABEIZAAAAgAiEMgAAAAARCGUAAAAAIhDKAAAAAEQglAEAAACIoDl2ByiD//7v//7ggw9i96I2rFq1KoSw2Wabxe4IVWTJkiUjRoyI3QuqyIYNG5YvX77NNtvE7ghVZPny5YMGDdpoo41id4Qq4u2DdlavXr1u3bpBgwbF7ghVxEDRTVtvvfX+++8fuxeVJZSpB++///7KlStj96I2zJ8/P51Ojxs3LnZHqCI/+9nPTj/99Ni9oIq89dZb99577z/+4z/G7ghV5M4779xzzz1HjhwZuyNUEW8ftPPUU0+tWrXqwAMPjN0Rqsh111134okn9u7dO3ZHatUmm2wSuwsVJ5SpB1OmTIndhZrRp0+ftra2adOmxe4IVWTu3LkzZ86M3QuqyIsvvvjGG2+oCrK98sorX/nKV/bee+/YHaGKePugnf/6r/9atmyZqiDbbbfdNn369H79+sXuCNXLPWUAAAAAIhDKAAAAAEQglAEAAACIQCgDAAAAEIFQBgAAACACoQwAAABABEIZAAAAgAiEMjSWpqampqamEEIqlUqlUkOGDGltbe0424YNG4YMGZLMk5nYt2/f7H9SN5KSCKqCj2X2spJoWB13fSqVyowVdj0JZxS04zyTjpxU0CmhDLXh/PPPL2ZIWrt2baqgI4444sgjj8zM/84778yfP79jO/Pnz3/nnXfKuQFUQJFVkXj77bdPOumkXXbZZcCAAWPGjDn77LM/+OCD5Km5c+dmz6kqaleXSiKdTv/mN7+ZNGnSlltuuemmm+63336//vWv0+l0CGH06NGzZ8/OzKkk6kw6nb7hhhv233//oUOHDho0aPz48VdeeeWGDRs6zpm9688666yxY8cmj+36WlT8fs8p5/CS/fZhoKhFpVVFgaUmT5583HHHZeZUFbWotKp4+eWXp0yZMnr06P79+3/qU5+aNWvWihUrkqeuvfbavn37ZuZUFeSQhqr3/vvvb7fddsWU69q1a/8ulxEjRoQQNt1008yc4eM/e55++ukd2zn77LObmpqS06/MxD59+jhkqkfxVZFOp19//fUhQ4aEEEaOHDlx4sShQ4eGELbbbrv33nsvezZVUdO6VBJtbW3Tp08PIfTr12/ChAnjxo3r3bt3COH888/Pnk1J1KVp06aFEPr373/AAQdMmDChf//+IYQJEyasX78+M49dX3+K2e/5FB5eVEvtKq0qjCH1rYSqeOyxx5LYZcyYMYcccsg222wTQhg6dOhbb72VPZuqIB/7mOrV2tr60ksv3XjjjXvttVc3M8QbbrghhPCrX/0qMyWE0KdPn/Hjx48ePbqtra3d/Lvuuuv48ePbjYOGxWpQWlUcccQRIYSzzjqrtbU1nU5/+OGHRx11VAjhpJNOyp5NVdSi0kri17/+dQhh5513XrRoUTLlueee23bbbfv27btixYrMbEqi/jz44INJPvvaa68lU5YuXZpcAvODH/wgM5tdX2eK3O/tFDm8qJYaVVpVGEPqWwlV0dbWts8++4QQfvGLXyRT1q5dO2XKlBDCCSeckD2nqiAf+5jqdccdd5Tlwq4lS5ZsttlmX/va17InJsPij370oxDCwoULs5969dVXQwg//OEPDYtVqISqaGtr23zzzfv3779u3brMxMWLF4cQRo8enT2nqqhFpQ0UEydODCHcc8892RNvuummEMKFF16YmaIk6s+sWbNCCNdee232xEceeSSE8NnPfjYzxa6vM0Xu93aKHF5US40qrSqMIfWthKpYvnx5CGGvvfbKzlmSvew8kyK5pwzVa6+99vp/H2tubi6tkXQ6PW3atM0333zOnDkdn/3yl78cQrj99tuzJ955550hhEMPPTRng6tXr541a9ZWW22V3ITisssua2trK61vlKCEqmhpaVm5cuXgwYOT76ckkm8wLV26tOP8qqK2lDZQPPvssyGEcePGZU/8whe+EEJI/kqWTUnUkwULFoQQxo8fnz1x9913DyG88sor7Wa26+tGl/Z7RpeGF9VSc0qrCmNIfSuhKl566aUQwtixY7NvO5XcOeH111/vOL+qIIfYqRAUpeSc+Oabbw4h3Hjjje2mhxD69OmTTqd32mmncePGZT910EEHJcF2zqx60qRJvXr12nvvvXfffffkBvuHHnpox0sQ6QHFV8XOO+8cQnjhhRcyU5I3v/Hjx2fPpipqXfElsdVWW4UQ3nnnneyJycnTmDFjMlOURP15+umn58+fv2HDhuyJDzzwQAjhc5/7XGaKXV9nitzvBRQYXlRLjSqtKowh9a2EqmhpaVm6dGm72xT+8Y9/DCHss88+2RNVBfkIZagNpYUy69at22GHHcaMGdNubE1nDYtnnnlmKpXK3Ihr1apVvXv3/va3v91xpck/t9hiiz/+8Y/JlEcffXTw4MEhhN/85jelbRfdUXxVJLcLGTVq1Ny5c5944omrrrpqyJAhW2211Z///Ofs2VRFrSu+JJLbDGW+/p24+OKLQwhDhgzJTFESjeDll1/eYYcdQgjXX399ZqJdX/dy7vcCigllVEut62pVFFhKVdSNLlXF4sWLH3rooTlz5owcOXKzzTZ77LHHsp9VFeQjlKE2lBbKXHHFFSGEu+66q+NvrFbpAAAK1UlEQVRTmWHxoYceCiFcc801yfRbbrklhPDAAw90XGnyzzlz5mS38+///u8hhIMPPrirfaP7iq+K1tbWCy64oN11gj/72c/apXWqotYVXxKvvfbapptuOnDgwF/+8pfvvPPOG2+8cfHFFyeL5wxllERdamtru+WWW5KfZpsxY0b2Xx3t+jpWYL8XUEwoo1pqV2lVYQypbyVURfI7EiGEpqamjiGOqiAfoQy1oYRQprW1dfjw4TvvvHPOMTQzLK5fv37w4MGHHXZYMn3q1KkDBw5MfvQu57CYuRl7YsmSJSGEkSNHdnWL6L4iq2L9+vXJd3TPOOOMl19+ec2aNS+99NL3v//9vn37HnDAAR1/aietKmpWlwaKl1566bDDDkt+wzKE8OlPfzr5SaacX19SEvXnueee23///UMIAwYM+OlPf5ovorXr60zh/V5AMaGMaqlRpVWFMaS+lTxWtLS0PProo3vssUcI4T//8z+zn1IV5COUoTaUEMokP5qQ/UMq2TLDYjqdnjp1ar9+/VpaWlpbW4cMGXLMMcfkXGnyz5aWlux21q9fH0Lo169fl/pGWRRZFdddd10IYfr06e2mX3vttSGE7J/lUhW1rrT0dsmSJWvWrEmn08m9+r70pS9lnlUSdWnDhg0/+MEPevfu3dzcfOKJJy5btqzjPHZ9/SlmvxdQTCiTVi21prSqMIbUt26OFYlnnnkmhLDTTjtlT1QV5OPXl6hbV111VVNT03HHHdfpnIceeuiHH3547733Pv744ytWrEhuip7Pm2++mf3PN954I4Sw/fbbd7O3VM5TTz0VQkh+AjnbYYcdFkK46667ci6lKurbunXr1q5du2HDhl69eg0fPrx///4hhOeffz6EkNwWuiMlUR/S6fSpp576z//8z9ttt90TTzxx5ZVXbr311oUXsevrQAn7vTSqpYaUVhXGkPpWwv696qqrDj/88ORsM2OXXXYJIfzlL3/Jt5SqIJtQhvq0bNmyu+66a8KECcOGDet05oMPPnijjTb67W9/e8cddzQ3N0+ePLnAzL/73e+y/3nbbbeFEHbbbbdudpjKGTVqVAjh1VdfbTf9hRdeCCEMHz4851Kqor5973vf69evX3LbqYzrr78+hDBt2rSciyiJ+nDrrbfOmTPni1/84pNPPjl27NhiFrHr60AJ+700qqWGlFYVxpD6VsL+bWtru/322//nf/4ne+Jrr70WQthuu+3yLaUq+ISo1+lAsQpcNnz//fc/+OCD7SZec801IYTvfe97+RoMWRcQptPpyZMnDx06dNddd/3CF76Qb6WZm4A++eSTyZSHH3544MCBIYTHH3+8hI2im4qsigULFvTu3XuTTTb5wx/+kJnhueee22mnnUIIs2fPzkxUFbWu+IHif//3f0MIw4YNW7x4cTLlyiuvDCG0+4lKJVF/Pv/5z4cQli5dWng2u77OFLnfc55RJIr8+lJatdSO0qrCGFLfSqiKF198sbm5eeDAgX/605+SKWvWrEmufDnllFOyl1IV5COUoTYUfzKUmDJlSghh3rx5+Rpst1TmD+aXXnppvpUm/zz00EObm5v33nvvsWPHNjU1hRDOOOOMEreK7im+Ki6//PJk/376058++OCDM39bOPzww1tbW/MtpSpqTpcGilmzZoUQ+vfvP2HChJEjR4YQNt988+eff77AUkqi1q1fv765uTmVSv1dLjvuuGNmTru+npS837MVP7yolppQWlUYQ+pbyfv3vPPOCyE0NTXtscceEydO3GKLLUIIo0aNWrVqVXb7qoJ8hDLUhi591mptbR04cGCvXr1Wr16dr8F2Sy1dujQZFl955ZV8K03+uWbNmlmzZm2xxRabb775xIkTb7rpptK3iu7pUlU88cQTX/nKV7bffvs+ffqMGjXqkEMOufPOO9v9MpeqqHVdKokNGzZcffXVY8eO7dev39ChQ4877rhFixYVXkpJ1LpFixaF/LL3tV1fT0re79mKH15US00orSqMIfWt5P3b1tZ2xx13JLdN6N+//5gxY84999z333+/XfuqgnxS6XS6QPEBAAAAUAlu9AsAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAEQhkAAACACIQyAAAAABEIZQAAAAAiEMoAAAAARCCUAQAAAIhAKAMAAAAQgVAGAAAAIAKhDAAAAEAE/x+YmJ2e9TicsQAAAABJRU5ErkJggg=="
    }
   },
   "cell_type": "markdown",
   "id": "66fd98c4",
   "metadata": {},
   "source": [
    "BedGraph is a file format that allows display of continuous-valued data in a track in genome browsers\n",
    "![bedgraphPlot.png](attachment:bedgraphPlot.png)\n",
    "pyGFT2bedGraph is the pyCRAC tool to generate bedgraphs. Inputs are the same as for the GTF2sgr, as it is a variation in the output of the same data.\n",
    "Parameters:\n",
    "* The --count flag instructs to include all reads (inc duplicates)\n",
    "* The --permillion flag instructs to normalise hits per million (the GTF file must contain the total number of mapped reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29cb445c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def makeCoverageBedgraphFiles(inputfile,outputfiles):\n",
    "\t\"\"\" takes the pyReadCounters gtf output files and generates bedgraph files for viewing of the data in genome browsers \"\"\"\n",
    "\toutputfile = re.search(\"(.*)_plus.bedgraph\",outputfiles[0]).group(1)\n",
    "\tcmd = \"pyGTF2bedGraph.py --gtf '%s' --count -v --permillion -o '%s' -c '%s'\" % (inputfile,outputfile,args.chromosome)\n",
    "\tlogger.info(cmd)\n",
    "\tos.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a0e48d7",
   "metadata": {},
   "source": [
    "Assessing the background level by negative controls is often not possible. An FDR algorithm can be used to indicate that the probability that the region being considered is not significantly enriched (i.e false positive).\n",
    "Input files: \n",
    "* (1) a pyReadCounters GTF output file\n",
    "* (2) a GTF annotation file\n",
    "* (3) a tab delimited file containing chromosome names and chromosome lengths.\n",
    "For each feature in the GTF annotation file it calculates the minimum number of over-lapping reads required to obtain an FDR less than or equal to the set threshold and reports genomic intervals with at least this read coverage in the GTF output file.\n",
    "Here they specify they only want to see peaks with suitable FDRs on protein-coding genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c19690c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def runPyCalculateFDRs(inputfile,outputfile):\n",
    "\t\"\"\" runs pyCalculateFDRs.py to find peaks with false discovery rates, on protein-coding genes \"\"\"\n",
    "\tcmd = \"pyCalculateFDRs.py -f '%s' -o '%s' -c '%s' --gtf '%s' -a protein_coding -m 0.01 --min=5 \" % (inputfile,outputfile,args.chromosome, args.gtf)\n",
    "\tlogger.info(cmd)\n",
    "\tos.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff92a54c",
   "metadata": {},
   "source": [
    "The genome browser is a very useful tool to visualise read distribution on chromosomes; however, often it is more convenient to generate separate pileup files for individual genes or chromosomal regions.\n",
    "input files: \n",
    "* (1) file containing the mapping data (-f option)\n",
    "* (2) a GTF feature file (- -gtf flag),\n",
    "* (3) a genomic reference sequence file in the tab format (--tab)\n",
    "* (4) a text file containing information about the genes/-transcripts of interest (-g).\n",
    "The code here will produce an output file for the provided genes (-g) that contains six columns in which each line consists of the gene or feature name, the 1-based coordinate, the reference base, the number of reads that cover that base, the number of times the reference base was substituted in reads and the number of times the reference base was deleted in reads."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05ce4a62",
   "metadata": {},
   "outputs": [],
   "source": [
    "def runPyPileup(inputfile,outputfile,alignfiletype = \"sam\"):\n",
    "\t\"\"\" runs pyPileup.py to make pileup tables of reads and deletions/mutations for a given genelist \"\"\"\n",
    "\tcmd = \"pyPileup.py -f '%s' -o '%s' --tab '%s' --gtf '%s' --file_type='%s' -g '%s'\" % (inputfile,outputfile,args.genometab, args.gtf, alignfiletype, args.genelist)\n",
    "\tlogger.info(cmd)\n",
    "\tos.system(cmd)\n",
    "\t"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e947b7bf",
   "metadata": {},
   "source": [
    "Given a file with aligned sequencing reads and a list of genomic features, a common task is to count how many reads map to each feature (a feature is here an interval, i.e.a range of positions, on a chromosome). In the case of RNA-Seq, the features are typically genes and each gene is considered as the union of all its exons. \n",
    "\n",
    "__The script htseq-count is a tool for RNA-Seq data analysis: Given a SAM/BAM file and a GTF or GFF file with gene models, it counts for each gene how many aligned reads overlap its exons.__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "364a5473",
   "metadata": {},
   "outputs": [],
   "source": [
    "def runHTSeqCountTranscripts(inputfile,outputfile,minaqual=2):\n",
    "\t\"\"\" runs htseq-count to assign reads to primary transcript features \"\"\"\n",
    "\t# Note: this did not play well with novoalign reads, so not being run for now.\n",
    "\tcmd = \"htseq-count -n 8 -a '%s' --counts_output '%s' '%s' '%s'  \" % (minaqual, outputfile, inputfile, args.gtf)\n",
    "\tlogger.info(cmd)\n",
    "\tos.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b020d46f",
   "metadata": {},
   "source": [
    "sort commands can be used to structure files (usually alphabeticaly) in a way that makes sense to the user. This function is organising and compressing .SAM files (generated in the pipeline) to .BAM format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa1cbdbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sortBamFiles(inputfile,outputfile):\n",
    "\t\"\"\" sorts .sam file, to produce a sorted compressed .bam \"\"\"\n",
    "\toutputfile = outputfile\n",
    "\tcmdsort = \"samtools view -b '%s' | samtools sort -@ 3 -O bam -o '%s' -T '%s'/tmp - \" % (inputfile, outputfile, root_dir)\n",
    "\tlogger.info(cmdsort)\n",
    "\tos.system(cmdsort)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61c1e719",
   "metadata": {},
   "source": [
    "The .sam files have now been compressed to the binary format .bam, a binary 'blob' that stores all of your aligned sequence data. Bam files can have a companion file, called an index file. This file has the same name, suffixed with .bai. This file acts like an external table of contents, and allows programs to jump directly to specific parts of the bam file without reading through all of the sequences. The below function acts to produce these, meaning an easily used library for the algorithm can be made."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1737276",
   "metadata": {},
   "outputs": [],
   "source": [
    "def indexBamFiles(inputfile,outputfile):\n",
    "\t\"\"\" indexes a sorted .bam file, to produce a matching .bam.bai \"\"\"\n",
    "\tcmdindex = \"samtools index '%s' \" % (inputfile)\n",
    "\tlogger.info(cmdindex)\n",
    "\tos.system(cmdindex)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afed9509",
   "metadata": {},
   "source": [
    "FastQC is a quality control tool for high throughput sequence data - provides a set of modular reports that you can use to assess if your data has any problems before you use it. Supports import of data from .BAM, .SAM and fastq files. Equally, BamQC is an application which takes a BAM file containing mapped data and runs a series of tests on it to generate a comprehensive QC report - judged on how far it departs from what you'd expect from a normal large data set with no significant biases."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d7761c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def runFastQC(inputfile,outputfile):\n",
    "\t\"\"\" runs fastQC on input fastq file \"\"\"\n",
    "\toutputdir = os.path.split(outputfile)[0]\n",
    "\tcmd = \"fastqc -o '%s' '%s'\" % (outputdir, inputfile)\n",
    "\tlogger.info(cmd)\n",
    "\tos.system(cmd)\n",
    "\n",
    "def runBamQC(inputfile,outputfile):\n",
    "\t\"\"\" runs BamQC on input bam/sam file \"\"\"\n",
    "\toutputdir = os.path.split(outputfile)[0]\n",
    "\tcmd = \"bamqc -o '%s' '%s'\" % (outputdir, inputfile)\n",
    "\tlogger.info(cmd)\n",
    "\tos.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8270f5f1",
   "metadata": {},
   "source": [
    "* BED files are a text file format used to store genomic regions as coordinates and associated annotations.\n",
    "bedtools multicov is a tool that reports the count (how many) of alignments in multiple indexed BAM (bai)files that are overlapping the intervals in a BED file. Specifically, for each BED interval provided, it reports a separate count of overlapping alignments from each BAM file.\n",
    "\n",
    "The early part of the code for this function is (I think) setting out the indexed BAM files to use the cmd= term calls multiBamCov, -s means require same strandedness (only report hits in B that overlap A on the same strand), -bed specifies the BED file with the region to be considered, -bams species the BAM files created above should be used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f5c763a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def runMultiCovTranscript(inputfiles,outputfile):\n",
    "\t\"\"\" counts reads to transcript features with bedtools multiBamCov \"\"\"\n",
    "\tinputfileorder = ' '.join(inputfiles)\n",
    "\tinputfilebasenames = [ os.path.basename(f) for f in inputfiles ] \n",
    "\tlogger.info(\"multiBamCov file order is %s\" % ' '.join(inputfilebasenames))\n",
    "\twith open(outputfile, \"w\") as outhandle:\n",
    "\t\touthandle.write( \"# allsample_transcriptcounts.txt - counts to transcripts from multiBamCov\\n\")\n",
    "\t\touthandle.write( \"# Run from CRAC_pipeline_SE_demult_dedup.py\\n\")\n",
    "\t\touthandle.write( \"# %s, date %s\\n\" % (args.name, time.strftime(\"%d-%m-%Y\")))\n",
    "\t\touthandle.write( \"%s\\n\" % '\\t'.join(GFF_FIELDS + inputfilebasenames))\n",
    "\tcmd = \"multiBamCov -s -bed '%s' -bams %s >> '%s'\" % (args.transcriptgff, inputfileorder, outputfile)\n",
    "\tlogger.info(cmd)\n",
    "\tos.system(cmd)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ac66a77",
   "metadata": {},
   "source": [
    "bedtools genomecov is a tool that produces summaries of feature coverage (e.g., aligned sequences) for a given genome. This includes in the bedGraph format, which allows display of continuous-valued data in track format as shades of colour/bars of varying height along the genome (see above).\n",
    "\n",
    "The cmd lines:\n",
    "* specify the genomeCoverageBed program\n",
    "* -bg asks the programe to report depth in BedGraph format (as histogram is the default)\n",
    "* -strand means calculate coverage of intervals from a specific strand (specified as + or -)\n",
    "* -ibam is the BAM file used as input for coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f271be03",
   "metadata": {},
   "outputs": [],
   "source": [
    "def makeGenomeCoverageBedgraph(inputfile,outputfiles):\n",
    "\t\"\"\" takes the sorted indexed bam files and generates bedgraph files for viewing of the data in genome browsers \"\"\"\n",
    "\toutputfile = re.search(\"(.*)_plus.bedgraph\",outputfiles[0]).group(1)\n",
    "\tcmdplus = \"genomeCoverageBed -bg -strand + -ibam %s > %s\" % (inputfile,outputfiles[0])\n",
    "\tlogger.info(cmdplus)\n",
    "\tos.system(cmdplus)\n",
    "\tcmdminus = \"genomeCoverageBed -bg -strand - -ibam %s > %s\" % (inputfile,outputfiles[1])\n",
    "\tlogger.info(cmdminus)\n",
    "\tos.system(cmdminus)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08e24c8c",
   "metadata": {},
   "source": [
    "To run the pipelune all of the dependent programs (including all of the pyCRAC tools used in functions above) must be installed. This function checks all of the required programs are installed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bb08b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "### checking if dependencies are installed\n",
    "\n",
    "dependencies = [\"flexbar\",\"pyGTF2sgr.py\",\"pyGTF2bedGraph.py\",\"pyReadCounters.py\",\"pyBarcodeFilter.py\",\"pyFastqJoiner.py\",\"pyFastqSplitter.py\",\"novoalign\",\"pyFastqDuplicateRemover.py\",\"samtools\",\"fastqc\",\"bamqc\",\"multiBamCov\"]\n",
    "\n",
    "for i in dependencies:\n",
    "\tcmd = \"where\" if platform.system() == \"Windows\" else \"which\"\n",
    "\ttry: \n",
    "\t\tsubprocess.call([cmd, \"%s\"])\n",
    "\texcept: \n",
    "\t\tsys.stderr.write(\"%s is not installed (properly) on your system. Please (re)install\" % i)\n",
    "\t\texit()\n",
    "\t\t"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27d174fe",
   "metadata": {},
   "source": [
    "Having specified all of the functions, the rufus pipeline can now be constructed. Rufus works to move through the pipeline as a series of files. You provide the python functions to provide the data transformations, tell ruffus how the functions are connected and it will direct the correct data through the pipeline at the correct time.\n",
    "\n",
    "NOTE: rufus requires python to be writen in a certain way so the coding here is very specific.\n",
    "\n",
    "The approximate steps in the CRAC analysis pipeline are:\n",
    "1. Trimming the reads - In the Illumina workflow adapter sequences are added to the end of RNA fragments by the PCR so that the cDNA has sequences complementary to graft sequences fixed on the flowcell, causing hybridisation. Bridge amplification can then generate clusters and forward strands can be sequenced (following denaturing and clevage). __The 3'adapter sequnce is one of the input files__\n",
    "2. Demultiplexing the reads - Multiplexing allows sequencing of libraries of RNA fragments derived from several samples in a single pool, as the number of reads produced from a single flow cell lane often exceeds the coverage required by a single library. To do this a short barcode of ~6nt is added to each DNA fragment during the PCR amplification step of library preparation. Following sequencing, this allows each read to be traced back to the original library and sorted into an appropriate ‘bin'. __The barcodes and the sample they correspond to are one of the input files__\n",
    "* The runfastQC tool is used to check the quality of the demultiplexed sequence data and the collapse duplicates function is used to remove any sequences considered to be PCR duplicates.\n",
    "3. Align the reads - The short reads generated by RNA-seq must be aligned to a reference genome. This provides biological information as aligned reads become annotated with the genomic features (e.g gene) in the region and the number of them approximates the abundance of those features in the original sample. Here, __Novoalign is the alignment algorithm used to do this__, and the function alignReads created earlier specified the parameters (match identically or penalties for insertions/deletions/gaps etc).\n",
    "* The quality control functions (runBAMQC) are used at this point to assess the quality of the alignment. The sorting and indexing tools are then used to structure the output BAM files.\n",
    "4. Mapping reads to genomic features - the functions used earlier are used to count (in tables) the number of reads in the Novoalign BAM outputs that are present within regions of particular genes (provided as an input).\n",
    "* The makeCoverageBedgraphFiles function is then used to convert the output of these counting functions (GTF and tables) into files that can be used by the BedGraphing algorithms\n",
    "5. Making BedGraph files - allows the information of the alignment (density at features) to be visualised along the genome\n",
    "* The pyPileUp function allows read distribution in a certain chromosomal region to be shown, while pyCalculateFDR can identify provide the probability that read density in a region is significantly enriched.\n",
    "* The bedtools functions defined earlier are used for presentation of the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82fbc74d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### start of pipeline\n",
    "\n",
    "# setting the main working directory\n",
    "home_dir = os.getcwd()\n",
    "root_dir = \"%s/%s\" % (home_dir,args.name)\n",
    "try:\n",
    "\tos.mkdir(root_dir)\n",
    "except OSError:\n",
    "\tpass\n",
    "\n",
    "### setting up logging\n",
    "if not args.log_file:\n",
    "\tlogfile = \"%s/%s\" % (root_dir,\"log_file.txt\")\n",
    "else:\n",
    "\tlogfile = \"%s/%s\" % (root_dir,args.log_file)\n",
    "logger, logger_mutex = cmdline.setup_logging (\"CRAC_Pipeline_SE\",logfile,10)\n",
    "\n",
    "### setting the starting files\n",
    "\n",
    "startingfiles = [os.path.abspath(i) for i in args.forwardreads]\n",
    "\n",
    "logger.info(\"analysing the following files:\\n%s\\n\" % \"\\n\".join(startingfiles))\n",
    "\n",
    "### start of pipeline definition\n",
    "\n",
    "pipeline = Pipeline(name=\"Single-end data CRAC pipeline for multiplexed libraries\")\n",
    "\n",
    "logger.info(\"Trimming the reads\")\n",
    "pipeline.transform(task_func = runFlexBar,\n",
    "\t\t\t\t\tinput  = startingfiles,\n",
    "\t\t\t\t\tfilter = formatter(\"^.+/([^/]+).(san)?fastq\"),\t\t\t\t\t\t\n",
    "\t\t\t\t\toutput = \"%s/flexbar_trimmed/{1[0]}_trimmed.fastq\" % root_dir,\n",
    "\t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"flexbar_trimmed\")))\n",
    "\t\t\t\n",
    "logger.info(\"Demultiplexing the reads\")\n",
    "pipeline.subdivide(task_func = demultiplexSamples,\n",
    "\t\t\t\t\tinput  = runFlexBar,\n",
    "\t\t\t\t\tfilter = formatter(\"^.+/([^/]+).fastq\"),\n",
    "\t\t\t\t\toutput = [\"%s/demultiplexed/{1[0]}_%s\" % (root_dir,barcodestring) for barcodestring in getBarcodeInfo(args.barcodes)]\n",
    "\t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"demultiplexed\")))\n",
    "\n",
    "#logger.info(\"demultiplexed fastq quality control\")\n",
    "pipeline.transform( task_func = runFastQC,\n",
    "\t\t\t\t\tinput  = demultiplexSamples,\n",
    "\t\t\t\t\tfilter = formatter(\"^.+/([^/]+).fastq\"),\n",
    "\t\t\t\t\toutput = \"%s/demultiplexed_fastqc/{1[0]}_fastqc.html\" % root_dir,\n",
    "\t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"demultiplexed_fastqc\")))\n",
    "\n",
    "pipeline.transform(task_func = collapseDuplicates,\n",
    "\t\t\t\t\tinput  = demultiplexSamples,\n",
    "\t\t\t\t\tfilter = formatter(\"^.+/([^/]+)_trimmed(?P<BARCODE>.*).fastq\"),\t\t\t\t\t\t\n",
    "\t\t\t\t\toutput = \"%s/collapsed/{1[0]}_trimmed{BARCODE[0]}.fasta\" % root_dir\n",
    "\t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"collapsed\")))\n",
    "\n",
    "logger.info(\"Aligning reads\")\n",
    "pipeline.transform( task_func = alignReads,\n",
    "\t\t\t\t\tinput  = collapseDuplicates,\n",
    "\t\t\t\t\tfilter = formatter(\"^.+/([^/]+).fasta\"),\n",
    "\t\t\t\t\toutput = \"%s/aligned_sam/{1[0]}.sam\" % root_dir,\n",
    "\t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"aligned_sam\")))\n",
    "\n",
    "#logger.info(\"aligned reads quality control\")\n",
    "pipeline.transform( task_func = runBamQC,\n",
    "\t\t\t\t\tinput  = alignReads,\n",
    "\t\t\t\t\tfilter = formatter(\"^.+/([^/]+).sam\"),\n",
    "\t\t\t\t\toutput = \"%s/aligned_bamqc/{1[0]}_bamqc.html\" % root_dir,\n",
    "\t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"aligned_bamqc\")))\n",
    "\n",
    "pipeline.transform( task_func = sortBamFiles,\n",
    "\t\t\t\t\tinput  = alignReads,\n",
    "\t\t\t\t\tfilter = formatter(\"^.+/([^/]+).sam\"),\n",
    "\t\t\t\t\toutput = \"%s/aligned_bamsorted/{1[0]}.bam\" % root_dir, \n",
    "\t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"aligned_bamsorted\")))\n",
    "\n",
    "pipeline.transform( task_func = indexBamFiles,\n",
    "\t\t\t\t\tinput  = sortBamFiles,\n",
    "\t\t\t\t\tfilter = formatter(\"^.+/([^/]+).bam\"),\n",
    "\t\t\t\t\toutput = \"%s/aligned_bamsorted/{1[0]}.bam.bai\" % root_dir, \n",
    "\t\t\t\t)\n",
    "\n",
    "logger.info(\"Mapping reads to genomic features\")\n",
    "pipeline.transform( task_func = runPyReadCounters,\n",
    "\t\t\t\t\tinput  = alignReads,\n",
    "\t\t\t\t\tfilter = formatter(\"^.+/([^/]+).sam\"),\n",
    "\t\t\t\t\toutput = \"%s/pyReadCounters_analyses/{1[0]}_count_output_reads.gtf\" % root_dir,\n",
    "\t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"pyReadCounters_analyses\"))).follows(alignReads)\n",
    "\n",
    "pipeline.transform(task_func = runPyReadCountersBlocksNoMuts,\n",
    "\t\t\t\t\tinput  = alignReads,\n",
    "\t\t\t\t\tfilter = formatter(\"^.+/([^/]+).sam\"),\n",
    "\t\t\t\t\toutput = \"%s/pyReadCountersBlocksNoMuts_analyses/{1[0]}_blocks_nomuts_count_output_cDNAs.gtf\" % root_dir,\n",
    "\t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"pyReadCountersBlocksNoMuts_analyses\"))).follows(alignReads)\n",
    "\n",
    "pipeline.subdivide(task_func = makeCoverageBedgraphFiles,\n",
    "\t\t\t\t\tinput  = runPyReadCountersBlocksNoMuts,\n",
    "\t\t\t\t\tfilter = formatter(\"^.+/([^/]+)_count_output_cDNAs.gtf\"),\n",
    "\t\t\t\t\toutput = [\"%s/bedgraph_files/{1[0]}_plus.bedgraph\" % root_dir,\"%s/bedgraph_files/{1[0]}_minus.bedgraph\" % root_dir]\n",
    "\t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"bedgraph_files\"))).follows(runPyReadCountersBlocksNoMuts)\n",
    "\n",
    "logger.info(\"Making bedgraph files\")\n",
    "\n",
    "## Add: pyPileUp.py to calculate pileups with and without mutations/deletions on select genes\n",
    "if args.genelist:\n",
    "\tpipeline.transform(task_func = runPyPileup,\n",
    "\t\t\t\t\t\tinput  = alignReads,\n",
    "\t\t\t\t\t\tfilter = formatter(\"^.+/([^/]+).sam\"),\n",
    "\t\t\t\t\t\toutput = \"%s/pyPileup_analyses/{1[0]}_pileups.txt\" % root_dir,\n",
    "\t\t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"pyPileup_analyses\"))).follows(alignReads)\n",
    "\n",
    "## Add: pyCalculateFDRs.py to calculate peaks and their occupancy\n",
    "pipeline.subdivide(task_func = runPyCalculateFDRs,\n",
    "\t\t\t\t\tinput  = runPyReadCounters,\n",
    "\t\t\t\t\tfilter = formatter(\"^.+/([^/]+)_count_output_reads.gtf\"),\n",
    "\t\t\t\t\toutput = \"%s/pyCalculateFDRs_analyses/{1[0]}_output_FDRs.gtf\" % root_dir\n",
    "\t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"pyCalculateFDRs_analyses\"))).follows(runPyReadCounters)\n",
    "\n",
    "## Add: something to add up the number of reads per feature. Maybe genes inc. UTRs. \n",
    "# htseq-count was giving weird results with most reads unaligned, and aligned reads not mapped to features.\n",
    "# placed on branch EW_SE_htseq and abandoned for now\n",
    "# pipeline.transform(task_func = runHTSeqCountTranscripts,\n",
    "# \t\t\t\t\tinput  = alignReads,\n",
    "# \t\t\t\t\tfilter = formatter(\"^.+/([^/]+).sam\"),\n",
    "# \t\t\t\t\toutput = \"%s/HTSeqCountTranscript_analysis/{1[0]}_transcript_counts.txt\" % root_dir,\n",
    "# \t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"HTSeqCountTranscript_analysis\"))).follows(alignReads)\n",
    "\n",
    "if args.transcriptgff:\n",
    "\tpipeline.merge(task_func = runMultiCovTranscript,\n",
    "\t\t\t\t\t\tinput  = sortBamFiles,\n",
    "\t\t\t\t\t\toutput = \"%s/multicov_analyses/allsample_transcriptcounts.txt\" % root_dir,\n",
    "\t\t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"multicov_analyses\"))).follows(indexBamFiles)\n",
    "\n",
    "pipeline.subdivide(task_func = makeGenomeCoverageBedgraph,\n",
    "\t\t\t\t\tinput  = sortBamFiles,\n",
    "\t\t\t\t\tfilter = formatter(\"^.+/([^/]+).bam\"),\n",
    "\t\t\t\t\toutput = [\"%s/bedgraph_genomecov/{1[0]}_plus.bedgraph\" % root_dir,\"%s/bedgraph_genomecov/{1[0]}_minus.bedgraph\" % root_dir]\n",
    "\t\t\t\t).follows(pipeline.mkdir(os.path.join(root_dir,\"bedgraph_genomecov\"))).follows(indexBamFiles)\n",
    "\n",
    "### print out flowchart describing pipeline\n",
    "pipeline_printout_graph(\"%s/flowchart.jpg\" %root_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c01b35de",
   "metadata": {},
   "source": [
    "The pipeline for 'Single-end data CRAC pipeline for multiplexed libraries' has now been constructed, specifiying the route through the functions for input and output files. This can then now be run :) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e1ba7d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### run the pipeline\n",
    "pipeline_run(multiprocess=args.processors,verbose=5,checksum_level=3)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
